// oblien/dist/runtime/exec.js
/**
 * Command execution inside the workspace runtime.
 *
 * ```ts
 * const result = await rt.exec.run(['echo', 'hello']);
 * console.log(result.stdout); // "hello\n"
 *
 * // Stream output
 * for await (const ev of rt.exec.stream(['npm', 'test'])) {
 *   if (ev.event === 'stdout') process.stdout.write(atob(ev.data));
 * }
 *
 * await rt.exec.kill('task_id');
 * ```
 */
export class ExecResource {
    _http;
    constructor(_http) {
        this._http = _http;
    }
    /** Run a command synchronously — returns when the process exits. */
    async run(cmd, params = {}) {
        return this._http.request({
            method: 'POST',
            path: '/exec',
            body: {
                cmd,
                exec_mode: params.execMode,
                timeout_seconds: params.timeoutSeconds,
                ttl_seconds: params.ttlSeconds,
                keep_logs: params.keepLogs,
            },
        });
    }
    /**
     * Stream command output via SSE — yields events as the process runs.
     *
     * Events: `task_id`, `stdout` (base64), `stderr` (base64), `exit`.
     *
     * ```ts
     * for await (const ev of rt.exec.stream(['npm', 'test'])) {
     *   if (ev.event === 'stdout') process.stdout.write(atob(ev.data));
     *   if (ev.event === 'exit') console.log('Exit code:', ev.exit_code);
     * }
     * ```
     */
    async *stream(cmd, params = {}) {
        const res = await this._http.requestStream({
            method: 'POST',
            path: '/exec/stream',
            body: {
                cmd,
                exec_mode: params.execMode,
                timeout_seconds: params.timeoutSeconds,
                ttl_seconds: params.ttlSeconds,
                keep_logs: params.keepLogs,
            },
        });
        if (!res.body)
            throw new Error('No response body for exec stream');
        yield* this.parseSSE(res);
    }
    /**
     * Subscribe to an existing task's output via SSE.
     * If the task already finished, replays output + exit immediately.
     *
     * ```ts
     * for await (const ev of rt.exec.subscribe('3')) {
     *   console.log(ev);
     * }
     * ```
     */
    async *subscribe(taskId) {
        const res = await this._http.requestStream({
            method: 'GET',
            path: '/exec/stream',
            query: { task_id: taskId },
        });
        if (!res.body)
            throw new Error('No response body for exec subscribe');
        yield* this.parseSSE(res);
    }
    /** List all tracked tasks. */
    async list() {
        const res = await this._http.request({
            method: 'GET',
            path: '/exec',
        });
        return res.tasks;
    }
    /** Get a specific task by ID. */
    async get(taskId) {
        return this._http.request({
            method: 'GET',
            path: `/exec/${taskId}`,
        });
    }
    /** Kill / remove a task. */
    async kill(taskId) {
        return this._http.request({
            method: 'DELETE',
            path: `/exec/${taskId}`,
        });
    }
    /** Delete all tasks. */
    async deleteAll() {
        return this._http.request({
            method: 'DELETE',
            path: '/exec',
        });
    }
    /** Send stdin to a running task. */
    async input(taskId, data) {
        return this._http.requestRaw(`/exec/${taskId}/input`, data);
    }
    // ── SSE Parser (shared by stream + subscribe) ───────────────────
    /**
     * Parse SSE named events from a Response stream.
     * Server sends `event: <name>\ndata: <json>\n\n`.
     */
    async *parseSSE(res) {
        const reader = res.body.getReader();
        const decoder = new TextDecoder();
        let buffer = '';
        let currentEvent = '';
        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done)
                    break;
                buffer += decoder.decode(value, { stream: true });
                const lines = buffer.split('\n');
                buffer = lines.pop() ?? '';
                for (const line of lines) {
                    if (line.startsWith('event: ')) {
                        currentEvent = line.slice(7).trim();
                    }
                    else if (line.startsWith('data: ')) {
                        const data = line.slice(6).trim();
                        if (!data || data === '[DONE]')
                            continue;
                        try {
                            const parsed = JSON.parse(data);
                            yield { event: currentEvent || 'message', ...parsed };
                        }
                        catch {
                            // Skip non-JSON data lines
                        }
                        currentEvent = '';
                    }
                    else if (line === '') {
                        // Empty line = event boundary, reset
                        currentEvent = '';
                    }
                }
            }
        }
        finally {
            reader.releaseLock();
        }
    }
}
