// oblien/dist/runtime/files.js
/**
 * File operations inside the workspace runtime.
 *
 * ```ts
 * const entries = await rt.files.list({ dirPath: '/app/src' });
 * const file = await rt.files.read({ filePath: '/app/index.js' });
 * await rt.files.write({ fullPath: '/app/hello.txt', content: 'hi' });
 * ```
 */
export class FilesResource {
    _http;
    constructor(_http) {
        this._http = _http;
    }
    /** List directory contents. */
    async list(params = {}) {
        return this._http.request({
            method: 'GET',
            path: '/files',
            query: {
                path: params.dirPath,
                nested: params.nested,
                flatten: params.flatten,
                light: params.light,
                include_hash: params.includeHash,
                include_content: params.includeContent,
                include_extensions: params.includeExtensions,
                code_files_only: params.codeFilesOnly,
                use_gitignore: params.useGitignore,
                max_depth: params.maxDepth,
                path_filter: params.pathFilter,
                include_ext: params.includeExt,
                ignore_patterns: params.ignorePatterns,
                max_content_budget: params.maxContentBudget,
            },
        });
    }
    /**
     * Stream directory entries as an async iterator (NDJSON).
     * Ideal for large directories — entries arrive as they're discovered.
     */
    async *stream(params = {}) {
        const res = await this._http.requestStream({
            method: 'GET',
            path: '/files/stream',
            query: {
                path: params.dirPath,
                nested: params.nested,
                flatten: params.flatten,
                light: params.light,
                include_hash: params.includeHash,
                include_content: params.includeContent,
                include_extensions: params.includeExtensions,
                code_files_only: params.codeFilesOnly,
                use_gitignore: params.useGitignore,
                max_depth: params.maxDepth,
                path_filter: params.pathFilter,
                include_ext: params.includeExt,
                ignore_patterns: params.ignorePatterns,
                max_content_budget: params.maxContentBudget,
            },
        });
        if (!res.body)
            throw new Error('No response body for file stream');
        const reader = res.body.getReader();
        const decoder = new TextDecoder();
        let buffer = '';
        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done)
                    break;
                buffer += decoder.decode(value, { stream: true });
                const lines = buffer.split('\n');
                buffer = lines.pop() ?? '';
                for (const line of lines) {
                    if (!line.trim())
                        continue;
                    const obj = JSON.parse(line);
                    // Skip control events (start, done)
                    if (obj.event)
                        continue;
                    yield obj;
                }
            }
        }
        finally {
            reader.releaseLock();
        }
    }
    /** Read file content. Supports line ranges. */
    async read(params) {
        return this._http.request({
            method: 'GET',
            path: '/files/read',
            query: {
                path: params.filePath,
                start_line: params.startLine,
                end_line: params.endLine,
                with_line_numbers: params.withLineNumbers,
            },
        });
    }
    /** Write or overwrite a file. */
    async write(params) {
        return this._http.request({
            method: 'POST',
            path: '/files/write',
            body: {
                path: params.fullPath,
                content: params.content,
                create_dirs: params.createDirs,
                append: params.append,
                mode: params.mode,
            },
        });
    }
    /** Create a directory (and parents). */
    async mkdir(params) {
        return this._http.request({
            method: 'POST',
            path: '/files/mkdir',
            body: { path: params.path, mode: params.mode },
        });
    }
    /** Get file/directory metadata. */
    async stat(params) {
        return this._http.request({
            method: 'GET',
            path: '/files/stat',
            query: { path: params.path },
        });
    }
    /** Delete a file or directory. Accepts path in body or query. */
    async delete(params) {
        return this._http.request({
            method: 'POST',
            path: '/files/delete',
            body: { path: params.path },
        });
    }
}
