// oblien/dist/runtime/search.js
/**
 * Content and filename search inside the workspace runtime.
 *
 * ```ts
 * const results = await rt.search.content({ query: 'handleRequest', path: '/app' });
 * const files = await rt.search.files({ query: 'controller' });
 * ```
 */
export class SearchResource {
    _http;
    constructor(_http) {
        this._http = _http;
    }
    /** Search file contents (powered by ripgrep). */
    async content(params) {
        return this._http.request({
            method: 'GET',
            path: '/files/search',
            query: {
                q: params.query,
                path: params.path,
                case_sensitive: params.caseSensitive,
                regex: params.regex,
                whole_word: params.wholeWord,
                max_results: params.maxResults,
                timeout: params.timeout,
                context_lines: params.contextLines,
                file_types: params.fileTypes,
                include_hidden: params.includeHidden,
                no_gitignore: params.noGitignore,
                ignore_patterns: params.ignorePatterns,
            },
        });
    }
    /** Search by filename/path. */
    async files(params) {
        return this._http.request({
            method: 'GET',
            path: '/files/search/files',
            query: {
                q: params.query,
                path: params.path,
                case_sensitive: params.caseSensitive,
                include_hidden: params.includeHidden,
                max_results: params.maxResults,
                ignore_patterns: params.ignorePatterns,
            },
        });
    }
    /** Check if ripgrep is installed. */
    async status() {
        return this._http.request({
            method: 'GET',
            path: '/files/search/init',
        });
    }
    /** Install ripgrep (one-time setup). */
    async init() {
        return this._http.request({
            method: 'POST',
            path: '/files/search/init',
        });
    }
}
