// oblien/dist/runtime/terminal.js
/**
 * Terminal sessions inside the workspace runtime.
 *
 * ```ts
 * const term = await rt.terminal.create({ shell: '/bin/bash', cols: 120, rows: 40 });
 * const sessions = await rt.terminal.list();
 * await rt.terminal.close(term.id);
 * ```
 */
export class TerminalResource {
    _http;
    constructor(_http) {
        this._http = _http;
    }
    /** Create a new terminal session. */
    async create(params = {}) {
        const body = {};
        if (params.shell)
            body.cmd = [params.shell];
        if (params.cmd)
            body.cmd = params.cmd;
        if (params.cols)
            body.cols = params.cols;
        if (params.rows)
            body.rows = params.rows;
        if (params.scrollbackSize)
            body.scrollback_size = params.scrollbackSize;
        return this._http.request({
            method: 'POST',
            path: '/terminals',
            body,
        });
    }
    /** List active terminal sessions. */
    async list() {
        const res = await this._http.request({
            method: 'GET',
            path: '/terminals',
        });
        return res.terminals;
    }
    /** Close a terminal session. */
    async close(sessionId) {
        return this._http.request({
            method: 'DELETE',
            path: `/terminals/${sessionId}`,
        });
    }
    /** Get scrollback buffer for a session. */
    async scrollback(sessionId, params) {
        return this._http.request({
            method: 'GET',
            path: `/terminals/${sessionId}/scrollback`,
            query: params?.bytes ? { bytes: params.bytes } : undefined,
        });
    }
}
