// oblien/dist/runtime/watcher.js
/**
 * File watcher management inside the workspace runtime.
 *
 * ```ts
 * const watcher = await rt.watcher.create({ path: '/app/src' });
 * const all = await rt.watcher.list();
 * await rt.watcher.delete(watcher.id);
 * ```
 */
export class WatcherResource {
    _http;
    constructor(_http) {
        this._http = _http;
    }
    /** Create a file watcher on a directory. */
    async create(params) {
        const res = await this._http.request({
            method: 'POST',
            path: '/watchers',
            body: {
                path: params.path,
                root: params.root,
                excludes: params.excludes,
                exclude: params.exclude,
            },
        });
        return res.watcher;
    }
    /** List active watchers. */
    async list() {
        const res = await this._http.request({
            method: 'GET',
            path: '/watchers',
        });
        return res.watchers;
    }
    /** Get watcher info by ID. */
    async get(watcherId) {
        const res = await this._http.request({
            method: 'GET',
            path: `/watchers/${watcherId}`,
        });
        return res.watcher;
    }
    /** Delete a watcher. */
    async delete(watcherId) {
        return this._http.request({
            method: 'DELETE',
            path: `/watchers/${watcherId}`,
        });
    }
}
