// oblien/dist/runtime/ws.js
/**
 * Multiplexed WebSocket connection for real-time terminal I/O and
 * file watcher events.
 *
 * Protocol (matches the guest server at /ws):
 *
 * **Binary frames** (terminal I/O):
 * - `[terminal_id_byte][raw data]` — both directions.
 * - Client → Server: stdin to terminal.
 * - Server → Client: stdout from terminal.
 *
 * **Text frames** (JSON with "channel" field):
 * - terminal: resize (client→server), exit (server→client)
 * - watcher: change, ready, overflow (server→client)
 *
 * Only one WS connection is active per VM — a new connection replaces the old one.
 *
 * ```ts
 * const ws = rt.ws();
 * ws.onTerminalOutput((id, data) => process.stdout.write(data));
 * ws.onWatcherEvent((ev) => console.log(ev.path, ev.op));
 * ws.connect();
 * ```
 */
export class RuntimeWebSocket {
    _ws = null;
    _baseUrl;
    _token;
    _reconnect;
    _reconnectDelay;
    _maxReconnectAttempts;
    _reconnectAttempts = 0;
    _intentionallyClosed = false;
    _reconnectTimer = null;
    // ── Event callbacks ──────────────────────────────────────────────
    /** Terminal stdout data: (terminalId, rawData). */
    _onTerminalOutput = null;
    /** Terminal exit event: (terminalId, exitCode). */
    _onTerminalExit = null;
    /** Watcher events (change, ready, overflow). */
    _onWatcherEvent = null;
    /** Connection opened. */
    _onOpen = null;
    /** Connection closed. */
    _onClose = null;
    /** Connection error. */
    _onError = null;
    /** Any text message (raw). */
    _onMessage = null;
    constructor(runtimeOptions, wsOptions) {
        const base = (runtimeOptions.baseUrl ?? 'https://workspace.oblien.com').replace(/\/$/, '');
        this._baseUrl = base.replace(/^http/, 'ws');
        this._token = runtimeOptions.token;
        this._reconnect = wsOptions?.reconnect ?? true;
        this._reconnectDelay = wsOptions?.reconnectDelay ?? 1000;
        this._maxReconnectAttempts = wsOptions?.maxReconnectAttempts ?? Infinity;
    }
    // ── Public API: Event Handlers ────────────────────────────────────
    /**
     * Register a handler for terminal output (binary frames).
     * `id` is the terminal session ID (string), `data` is the raw output.
     */
    onTerminalOutput(handler) {
        this._onTerminalOutput = handler;
        return this;
    }
    /** Register a handler for terminal exit events. */
    onTerminalExit(handler) {
        this._onTerminalExit = handler;
        return this;
    }
    /** Register a handler for file watcher events (change, ready, overflow). */
    onWatcherEvent(handler) {
        this._onWatcherEvent = handler;
        return this;
    }
    /** Register a handler called when the WebSocket connects. */
    onOpen(handler) {
        this._onOpen = handler;
        return this;
    }
    /** Register a handler called when the WebSocket closes. */
    onClose(handler) {
        this._onClose = handler;
        return this;
    }
    /** Register a handler for errors. */
    onError(handler) {
        this._onError = handler;
        return this;
    }
    /** Register a handler for all incoming text messages (raw parsed JSON). */
    onMessage(handler) {
        this._onMessage = handler;
        return this;
    }
    // ── Public API: Send ──────────────────────────────────────────────
    /**
     * Write stdin data to a terminal session.
     * Binary frame format: `[terminal_id_byte][data]`.
     *
     * `terminalId` is the numeric session ID (same as `parseInt(session.id)`).
     * Accepts either a `string` (e.g. `"1"`) or `number`.
     */
    writeTerminalInput(terminalId, data) {
        if (!this._ws || this._ws.readyState !== WebSocket.OPEN)
            return;
        const id = typeof terminalId === 'string' ? parseInt(terminalId, 10) : terminalId;
        const payload = typeof data === 'string' ? new TextEncoder().encode(data) : data;
        const frame = new Uint8Array(1 + payload.length);
        frame[0] = id;
        frame.set(payload, 1);
        this._ws.send(frame);
    }
    /** Send a terminal resize command. */
    resizeTerminal(terminalId, cols, rows) {
        this.sendJSON({
            channel: 'terminal',
            type: 'resize',
            id: terminalId,
            cols,
            rows,
        });
    }
    /** Send a raw JSON text message. */
    sendJSON(msg) {
        if (!this._ws || this._ws.readyState !== WebSocket.OPEN)
            return;
        this._ws.send(JSON.stringify(msg));
    }
    // ── Public API: Lifecycle ─────────────────────────────────────────
    /** Open the WebSocket connection. */
    connect() {
        this._intentionallyClosed = false;
        this._reconnectAttempts = 0;
        this._doConnect();
    }
    /** Close the WebSocket connection (no reconnect). */
    close() {
        this._intentionallyClosed = true;
        if (this._reconnectTimer) {
            clearTimeout(this._reconnectTimer);
            this._reconnectTimer = null;
        }
        if (this._ws) {
            this._ws.close(1000, 'client close');
            this._ws = null;
        }
    }
    /** Whether the socket is currently open. */
    get connected() {
        return this._ws?.readyState === WebSocket.OPEN;
    }
    // ── Internals ─────────────────────────────────────────────────────
    _doConnect() {
        const url = `${this._baseUrl}/ws?token=${encodeURIComponent(this._token)}`;
        const ws = new WebSocket(url);
        ws.binaryType = 'arraybuffer';
        this._ws = ws;
        ws.addEventListener('open', () => {
            this._reconnectAttempts = 0;
            this._onOpen?.();
        });
        ws.addEventListener('close', (ev) => {
            this._onClose?.(ev.code, ev.reason);
            this._ws = null;
            if (!this._intentionallyClosed && this._reconnect) {
                this._scheduleReconnect();
            }
        });
        ws.addEventListener('error', (ev) => {
            this._onError?.(ev);
        });
        ws.addEventListener('message', (ev) => {
            if (ev.data instanceof ArrayBuffer) {
                this._handleBinary(new Uint8Array(ev.data));
            }
            else if (typeof ev.data === 'string') {
                this._handleText(ev.data);
            }
        });
    }
    _handleBinary(frame) {
        if (frame.length < 2)
            return;
        const id = String(frame[0]);
        const data = frame.subarray(1);
        this._onTerminalOutput?.(id, data);
    }
    _handleText(raw) {
        let msg;
        try {
            msg = JSON.parse(raw);
        }
        catch {
            return;
        }
        this._onMessage?.(msg);
        const channel = msg.channel || 'terminal';
        switch (channel) {
            case 'terminal':
                this._handleTerminalMessage(msg);
                break;
            case 'watcher':
                this._onWatcherEvent?.(msg);
                break;
        }
    }
    _handleTerminalMessage(msg) {
        const type = msg.type;
        if (type === 'exit') {
            const ev = msg;
            this._onTerminalExit?.(ev.id, ev.code);
        }
    }
    _scheduleReconnect() {
        if (this._reconnectAttempts >= this._maxReconnectAttempts)
            return;
        this._reconnectAttempts++;
        // Exponential backoff capped at 30s
        const delay = Math.min(this._reconnectDelay * Math.pow(1.5, this._reconnectAttempts - 1), 30_000);
        this._reconnectTimer = setTimeout(() => {
            this._reconnectTimer = null;
            this._doConnect();
        }, delay);
    }
}
