// oblien/dist/cli/output.js
/**
 * Output utilities — JSON by default, table when --table is passed.
 */
export function printJson(data) {
    console.log(JSON.stringify(data, null, 2));
}
export function printTable(rows, columns) {
    if (rows.length === 0) {
        console.log('(no results)');
        return;
    }
    const cols = columns ?? Object.keys(rows[0]);
    // Calculate column widths
    const widths = cols.map(col => Math.max(col.length, ...rows.map(r => String(r[col] ?? '').length)));
    // Header
    const header = cols.map((col, i) => col.padEnd(widths[i])).join('  ');
    const separator = widths.map(w => '─'.repeat(w)).join('──');
    console.log(header);
    console.log(separator);
    for (const row of rows) {
        const line = cols.map((col, i) => String(row[col] ?? '').padEnd(widths[i])).join('  ');
        console.log(line);
    }
}
export function printSuccess(message) {
    console.log(`✓ ${message}`);
}
export function printError(message) {
    console.error(`✗ ${message}`);
}
export function printKeyValue(data, indent = 0) {
    const pad = ' '.repeat(indent);
    for (const [key, value] of Object.entries(data)) {
        if (value === undefined || value === null)
            continue;
        if (typeof value === 'object' && !Array.isArray(value)) {
            console.log(`${pad}${key}:`);
            printKeyValue(value, indent + 2);
        }
        else if (Array.isArray(value)) {
            console.log(`${pad}${key}: ${value.join(', ')}`);
        }
        else {
            console.log(`${pad}${key}: ${value}`);
        }
    }
}
/**
 * Parse global flags from args, return remaining args.
 */
export function extractGlobalFlags(args) {
    const result = {
        json: false,
        clientId: undefined,
        clientSecret: undefined,
        baseUrl: undefined,
        remaining: [],
    };
    let i = 0;
    while (i < args.length) {
        if (args[i] === '--json') {
            result.json = true;
            i++;
        }
        else if (args[i] === '--client-id' && i + 1 < args.length) {
            result.clientId = args[++i];
            i++;
        }
        else if (args[i] === '--client-secret' && i + 1 < args.length) {
            result.clientSecret = args[++i];
            i++;
        }
        else if (args[i] === '--base-url' && i + 1 < args.length) {
            result.baseUrl = args[++i];
            i++;
        }
        else {
            result.remaining.push(args[i]);
            i++;
        }
    }
    return result;
}
/** Print a result — JSON if --json flag, otherwise key-value. */
export function printResult(data, json) {
    if (json) {
        printJson(data);
    }
    else if (Array.isArray(data)) {
        if (data.length === 0) {
            console.log('(no results)');
        }
        else if (typeof data[0] === 'object') {
            printTable(data);
        }
        else {
            for (const item of data)
                console.log(item);
        }
    }
    else if (typeof data === 'object' && data !== null) {
        printKeyValue(data);
    }
    else {
        console.log(data);
    }
}
