// oblien/dist/cli/commands/api-access.js
import { buildClient } from '../auth.js';
import { printResult, printSuccess } from '../output.js';
export async function apiAccessCmd(args, flags) {
    const sub = args[0];
    switch (sub) {
        case 'enable': return aaEnable(args[1], flags);
        case 'disable': return aaDisable(args[1], flags);
        case 'status': return aaStatus(args[1], flags);
        case 'token': return aaToken(args[1], flags);
        case 'raw-token': return aaRawToken(args[1], flags);
        case 'rotate': return aaRotate(args[1], flags);
        default:
            console.log(`
  oblien api-access — Manage Internal API access

  Usage:
    oblien api-access enable <ws-id>       Enable Internal API (returns JWT)
    oblien api-access disable <ws-id>      Disable Internal API
    oblien api-access status <ws-id>       Check status
    oblien api-access token <ws-id>        Get gateway JWT
    oblien api-access raw-token <ws-id>    Get raw connection token + IP
    oblien api-access rotate <ws-id>       Rotate token

  Examples:
    oblien api-access enable ws_abc123
    oblien api-access token ws_abc123 --json
    oblien api-access raw-token ws_abc123
`);
    }
}
function requireId(id, cmd) {
    if (!id || id.startsWith('-')) {
        console.error(`Usage: oblien api-access ${cmd} <workspace-id>`);
        process.exit(1);
    }
    return id;
}
async function aaEnable(id, flags) {
    const wsId = requireId(id, 'enable');
    const client = buildClient(flags);
    const res = await client.workspaces.apiAccess.enable(wsId);
    if (flags.json) {
        printResult(res, true);
    }
    else {
        printSuccess(`Internal API enabled for ${wsId}`);
        if (res.token) {
            console.log(`\nGateway JWT:\n${res.token}\n`);
        }
    }
}
async function aaDisable(id, flags) {
    const wsId = requireId(id, 'disable');
    const client = buildClient(flags);
    await client.workspaces.apiAccess.disable(wsId);
    printSuccess(`Internal API disabled for ${wsId}`);
}
async function aaStatus(id, flags) {
    const wsId = requireId(id, 'status');
    const client = buildClient(flags);
    const res = await client.workspaces.apiAccess.status(wsId);
    printResult(res, flags.json);
}
async function aaToken(id, flags) {
    const wsId = requireId(id, 'token');
    const client = buildClient(flags);
    const res = await client.workspaces.apiAccess.getToken(wsId);
    if (flags.json) {
        printResult(res, true);
    }
    else {
        console.log(res.token ?? JSON.stringify(res));
    }
}
async function aaRawToken(id, flags) {
    const wsId = requireId(id, 'raw-token');
    const client = buildClient(flags);
    const res = await client.workspaces.apiAccess.rawToken(wsId);
    printResult(res, flags.json);
}
async function aaRotate(id, flags) {
    const wsId = requireId(id, 'rotate');
    const client = buildClient(flags);
    const res = await client.workspaces.apiAccess.rotateToken(wsId);
    printResult(res, flags.json);
}
