// oblien/dist/cli/commands/config.js
import { saveConfig, loadConfig, clearConfig, configPath } from '../config-store.js';
import { printSuccess, printKeyValue } from '../output.js';
export async function configCmd(args, _flags) {
    const sub = args[0];
    switch (sub) {
        case 'set':
            return configSet(args.slice(1));
        case 'show':
            return configShow();
        case 'path':
            console.log(configPath());
            return;
        case 'clear':
            clearConfig();
            printSuccess('Credentials cleared.');
            return;
        default:
            console.log(`
  oblien config — Manage credentials

  Usage:
    oblien config set --client-id <id> --client-secret <secret>
    oblien config show
    oblien config path
    oblien config clear

  Credentials are stored in ${configPath()}
`);
    }
}
function configSet(args) {
    let clientId;
    let clientSecret;
    let baseUrl;
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--client-id' && i + 1 < args.length)
            clientId = args[++i];
        else if (args[i] === '--client-secret' && i + 1 < args.length)
            clientSecret = args[++i];
        else if (args[i] === '--base-url' && i + 1 < args.length)
            baseUrl = args[++i];
    }
    if (!clientId || !clientSecret) {
        console.error('Usage: oblien config set --client-id <id> --client-secret <secret> [--base-url <url>]');
        process.exit(1);
    }
    saveConfig({ clientId, clientSecret, baseUrl });
    printSuccess(`Credentials saved to ${configPath()}`);
}
function configShow() {
    const config = loadConfig();
    if (!config || !config.clientId) {
        console.log('No credentials configured.\n\nRun: oblien config set --client-id <id> --client-secret <secret>');
        return;
    }
    printKeyValue({
        clientId: config.clientId,
        clientSecret: config.clientSecret.slice(0, 8) + '…',
        baseUrl: config.baseUrl ?? 'https://api.oblien.com (default)',
        path: configPath(),
    });
}
