// oblien/dist/cli/commands/exec.js
import { buildClient } from '../auth.js';
import { printResult } from '../output.js';
export async function execCmd(rawArgs, flags) {
    // Parse: oblien exec <ws-id> [--stream] [--json] -- <cmd...>
    // Or:    oblien exec <ws-id> <cmd...>
    // Strip global flags first
    const args = rawArgs.filter((a, i) => {
        if (a === '--json')
            return false;
        if (a === '--client-id' || a === '--client-secret' || a === '--base-url') {
            rawArgs[i + 1] = ''; // mark next as consumed
            return false;
        }
        return a !== '';
    });
    const wsId = args[0];
    // Subcommands that don't run a command
    if (wsId === 'list')
        return execList(args[1], flags);
    if (wsId === 'get')
        return execGet(args[1], args[2], flags);
    if (wsId === 'kill')
        return execKill(args[1], args[2], flags);
    if (wsId === 'delete-all')
        return execDeleteAll(args[1], flags);
    if (wsId === 'input')
        return execInput(args[1], args[2], args.slice(3), flags);
    if (wsId === 'subscribe')
        return execSubscribe(args[1], args[2], flags);
    if (!wsId || wsId.startsWith('-')) {
        console.log(`
  oblien exec — Execute commands in a workspace

  Usage:
    oblien exec <workspace-id> -- <command...>       Run a command
    oblien exec <workspace-id> --stream -- <cmd>     Stream output (SSE)
    oblien exec list <ws-id>                         List running tasks
    oblien exec get <ws-id> <task-id>                Get task info
    oblien exec kill <ws-id> <task-id>               Kill a task
    oblien exec delete-all <ws-id>                   Delete all tasks
    oblien exec input <ws-id> <task-id> <data>       Send stdin to task
    oblien exec subscribe <ws-id> <task-id>          Stream task output

  Examples:
    oblien exec ws_abc123 -- npm test
    oblien exec ws_abc123 -- ls -la /app
    oblien exec ws_abc123 --stream -- npm run dev
    oblien exec list ws_abc123
    oblien exec kill ws_abc123 task_def456
    oblien exec subscribe ws_abc123 task_def456
`);
        return;
    }
    const stream = args.includes('--stream');
    // Find everything after --
    const dashIdx = args.indexOf('--');
    let cmd;
    if (dashIdx >= 0) {
        cmd = args.slice(dashIdx + 1);
    }
    else {
        // No -- separator, take everything after wsId and flags
        cmd = args.slice(1).filter(a => a !== '--stream');
    }
    if (cmd.length === 0) {
        console.error('Error: No command specified.\n\nUsage: oblien exec <ws-id> -- <command...>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    if (stream) {
        for await (const event of rt.exec.stream(cmd)) {
            if ('data' in event) {
                process.stdout.write(String(event.data));
            }
            else if (flags.json) {
                printResult(event, true);
            }
        }
    }
    else {
        const result = await rt.exec.run(cmd);
        printResult(result, flags.json);
    }
}
async function execList(wsId, flags) {
    if (!wsId || wsId.startsWith('-')) {
        console.error('Usage: oblien exec list <ws-id>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const res = await rt.exec.list();
    printResult(res, flags.json);
}
async function execGet(wsId, taskId, flags) {
    if (!wsId) {
        console.error('Usage: oblien exec get <ws-id> <task-id>');
        process.exit(1);
    }
    if (!taskId) {
        console.error('Usage: oblien exec get <ws-id> <task-id>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const res = await rt.exec.get(taskId);
    printResult(res, flags.json);
}
async function execKill(wsId, taskId, flags) {
    if (!wsId) {
        console.error('Usage: oblien exec kill <ws-id> <task-id>');
        process.exit(1);
    }
    if (!taskId) {
        console.error('Usage: oblien exec kill <ws-id> <task-id>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const res = await rt.exec.kill(taskId);
    printResult(res, flags.json);
}
async function execDeleteAll(wsId, flags) {
    if (!wsId || wsId.startsWith('-')) {
        console.error('Usage: oblien exec delete-all <ws-id>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const res = await rt.exec.deleteAll();
    printResult(res, flags.json);
}
async function execInput(wsId, taskId, rest, flags) {
    if (!wsId) {
        console.error('Usage: oblien exec input <ws-id> <task-id> <data>');
        process.exit(1);
    }
    if (!taskId) {
        console.error('Usage: oblien exec input <ws-id> <task-id> <data>');
        process.exit(1);
    }
    const data = rest.join(' ');
    if (!data) {
        console.error('Usage: oblien exec input <ws-id> <task-id> <data>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const res = await rt.exec.input(taskId, data);
    printResult(res, flags.json);
}
async function execSubscribe(wsId, taskId, flags) {
    if (!wsId) {
        console.error('Usage: oblien exec subscribe <ws-id> <task-id>');
        process.exit(1);
    }
    if (!taskId) {
        console.error('Usage: oblien exec subscribe <ws-id> <task-id>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    for await (const event of rt.exec.subscribe(taskId)) {
        if ('data' in event) {
            process.stdout.write(String(event.data));
        }
        else if (flags.json) {
            printResult(event, true);
        }
    }
}
