// oblien/dist/cli/commands/files.js
import { readFileSync } from 'node:fs';
import { buildClient } from '../auth.js';
import { printResult, printTable, printSuccess } from '../output.js';
export async function filesCmd(args, flags) {
    const sub = args[0];
    switch (sub) {
        case 'ls':
        case 'list': return filesList(args[1], args.slice(2), flags);
        case 'stream-ls': return filesStreamLs(args[1], args.slice(2), flags);
        case 'cat':
        case 'read': return filesRead(args[1], args[2], flags);
        case 'write': return filesWrite(args[1], args[2], args.slice(3), flags);
        case 'rm':
        case 'delete': return filesDelete(args[1], args[2], flags);
        case 'stat': return filesStat(args[1], args[2], flags);
        case 'mkdir': return filesMkdir(args[1], args[2], flags);
        default:
            console.log(`
  oblien files — Manage workspace files

  Usage:
    oblien files ls <ws-id> [path]                   List files
    oblien files stream-ls <ws-id> [path]              Stream file listing (NDJSON)
    oblien files cat <ws-id> <path>                  Read a file
    oblien files write <ws-id> <path> [--content <text> | --file <local-path>]
                                                     Write a file
    oblien files rm <ws-id> <path>                   Delete file/directory
    oblien files stat <ws-id> <path>                 Get file info
    oblien files mkdir <ws-id> <path>                Create directory

  Examples:
    oblien files ls ws_abc123 /app
    oblien files cat ws_abc123 /app/index.js
    oblien files write ws_abc123 /app/config.json --content '{"port": 3000}'
    oblien files write ws_abc123 /app/main.py --file ./main.py
    echo "hello" | oblien files write ws_abc123 /app/hello.txt
    oblien files rm ws_abc123 /app/tmp
    oblien files mkdir ws_abc123 /app/logs
`);
    }
}
function requireArgs(wsId, cmd) {
    if (!wsId || wsId.startsWith('-')) {
        console.error(`Usage: oblien files ${cmd} <workspace-id> ...`);
        process.exit(1);
    }
    return wsId;
}
async function filesList(wsId, args, flags) {
    const id = requireArgs(wsId, 'ls');
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(id);
    const dirPath = args[0] || '/';
    const res = await rt.files.list({ dirPath });
    if (flags.json) {
        printResult(res, true);
    }
    else {
        const entries = res.entries ?? res;
        if (Array.isArray(entries)) {
            printTable(entries.map((f) => ({
                name: f.name + (f.type === 'directory' ? '/' : ''),
                size: f.size ?? '-',
                modified: f.modified?.slice(0, 19) ?? '-',
            })), ['name', 'size', 'modified']);
        }
        else {
            printResult(res, false);
        }
    }
}
async function filesRead(wsId, filePath, flags) {
    const id = requireArgs(wsId, 'cat');
    if (!filePath) {
        console.error('Usage: oblien files cat <ws-id> <file-path>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(id);
    const res = await rt.files.read({ filePath });
    if (flags.json) {
        printResult(res, true);
    }
    else {
        // Print raw content for piping
        process.stdout.write(res.content ?? JSON.stringify(res));
    }
}
async function filesWrite(wsId, filePath, args, flags) {
    const id = requireArgs(wsId, 'write');
    if (!filePath) {
        console.error('Usage: oblien files write <ws-id> <path> [--content <text> | --file <local>]');
        process.exit(1);
    }
    let content;
    const contentFlag = args.indexOf('--content');
    const fileFlag = args.indexOf('--file');
    if (contentFlag >= 0 && contentFlag + 1 < args.length) {
        content = args[contentFlag + 1];
    }
    else if (fileFlag >= 0 && fileFlag + 1 < args.length) {
        content = readFileSync(args[fileFlag + 1], 'utf-8');
    }
    else {
        // Read from stdin
        const chunks = [];
        for await (const chunk of process.stdin) {
            chunks.push(chunk);
        }
        content = Buffer.concat(chunks).toString('utf-8');
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(id);
    const res = await rt.files.write({ fullPath: filePath, content, createDirs: true });
    printResult(res, flags.json);
}
async function filesDelete(wsId, filePath, flags) {
    const id = requireArgs(wsId, 'rm');
    if (!filePath) {
        console.error('Usage: oblien files rm <ws-id> <path>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(id);
    await rt.files.delete({ path: filePath });
    printSuccess(`Deleted: ${filePath}`);
}
async function filesStat(wsId, filePath, flags) {
    const id = requireArgs(wsId, 'stat');
    if (!filePath) {
        console.error('Usage: oblien files stat <ws-id> <path>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(id);
    const res = await rt.files.stat({ path: filePath });
    printResult(res, flags.json);
}
async function filesMkdir(wsId, dirPath, flags) {
    const id = requireArgs(wsId, 'mkdir');
    if (!dirPath) {
        console.error('Usage: oblien files mkdir <ws-id> <path>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(id);
    await rt.files.mkdir({ path: dirPath });
    printSuccess(`Created: ${dirPath}`);
}
async function filesStreamLs(wsId, args, flags) {
    const id = requireArgs(wsId, 'stream-ls');
    const dirPath = args.find(a => !a.startsWith('-')) ?? '/';
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(id);
    const params = { path: dirPath };
    if (args.includes('--recursive'))
        params.recursive = true;
    for await (const entry of rt.files.stream(params)) {
        if (flags.json) {
            printResult(entry, true);
        }
        else {
            const name = entry.name ?? entry.path ?? JSON.stringify(entry);
            process.stdout.write(`${name}\n`);
        }
    }
}
