// oblien/dist/cli/commands/lifecycle.js
import { buildClient } from '../auth.js';
import { printResult, printSuccess } from '../output.js';
export async function lifecycleCmd(args, flags) {
    const sub = args[0];
    switch (sub) {
        case 'get': return lcGet(args[1], flags);
        case 'make-permanent': return lcMakePermanent(args[1], flags);
        case 'make-temporary': return lcMakeTemporary(args[1], args.slice(2), flags);
        case 'update-ttl': return lcUpdateTtl(args[1], args.slice(2), flags);
        case 'destroy': return lcDestroy(args[1], flags);
        default:
            console.log(`
  oblien lifecycle — Manage workspace lifecycle mode

  Usage:
    oblien lifecycle get <ws-id>                        Get lifecycle status
    oblien lifecycle make-permanent <ws-id>              Set permanent mode
    oblien lifecycle make-temporary <ws-id> [flags]      Set temporary mode
    oblien lifecycle update-ttl <ws-id> [flags]          Update TTL
    oblien lifecycle destroy <ws-id>                     Destroy workspace

  Temporary mode flags:
    --ttl <seconds>          Time-to-live in seconds
    --ttl-action <action>    stop or remove (default: stop)

  Examples:
    oblien lifecycle get ws_abc123
    oblien lifecycle make-temporary ws_abc123 --ttl 3600
    oblien lifecycle make-permanent ws_abc123
`);
    }
}
function requireId(id, cmd) {
    if (!id || id.startsWith('-')) {
        console.error(`Usage: oblien lifecycle ${cmd} <workspace-id>`);
        process.exit(1);
    }
    return id;
}
function parseFlag(args, flag) {
    const idx = args.indexOf(flag);
    return idx >= 0 && idx + 1 < args.length ? args[idx + 1] : undefined;
}
async function lcGet(id, flags) {
    const wsId = requireId(id, 'get');
    const client = buildClient(flags);
    const res = await client.workspaces.lifecycle.get(wsId);
    printResult(res, flags.json);
}
async function lcMakePermanent(id, flags) {
    const wsId = requireId(id, 'make-permanent');
    const client = buildClient(flags);
    await client.workspaces.lifecycle.makePermanent(wsId);
    printSuccess(`Set ${wsId} to permanent mode`);
}
async function lcMakeTemporary(id, args, flags) {
    const wsId = requireId(id, 'make-temporary');
    const client = buildClient(flags);
    const params = {};
    const ttl = parseFlag(args, '--ttl');
    const action = parseFlag(args, '--ttl-action');
    if (ttl)
        params.ttl = Number(ttl);
    if (action)
        params.ttl_action = action;
    await client.workspaces.lifecycle.makeTemporary(wsId, params);
    printSuccess(`Set ${wsId} to temporary mode${ttl ? ` (TTL: ${ttl}s)` : ''}`);
}
async function lcUpdateTtl(id, args, flags) {
    const wsId = requireId(id, 'update-ttl');
    const client = buildClient(flags);
    const params = {};
    const ttl = parseFlag(args, '--ttl');
    const action = parseFlag(args, '--ttl-action');
    if (ttl)
        params.ttl = Number(ttl);
    if (action)
        params.ttl_action = action;
    await client.workspaces.lifecycle.updateTtl(wsId, params);
    printSuccess(`Updated TTL for ${wsId}`);
}
async function lcDestroy(id, flags) {
    const wsId = requireId(id, 'destroy');
    const client = buildClient(flags);
    await client.workspaces.lifecycle.destroy(wsId);
    printSuccess(`Destroyed ${wsId}`);
}
