// oblien/dist/cli/commands/logs.js
import { buildClient } from '../auth.js';
import { printResult } from '../output.js';
export async function logsCmd(args, flags) {
    const sub = args[0];
    switch (sub) {
        case 'get': return logsGet(args[1], args.slice(2), flags);
        case 'boot': return logsStream('boot', args[1], args.slice(2), flags);
        case 'cmd': return logsStream('cmd', args[1], args.slice(2), flags);
        case 'files': return logsFiles(args[1], flags);
        case 'file': return logsFile(args[1], args[2], flags);
        case 'clear': return logsClear(args[1], flags);
        default:
            // If first arg looks like a workspace ID, treat as `logs get`
            if (sub && !sub.startsWith('-')) {
                return logsGet(sub, args.slice(1), flags);
            }
            console.log(`
  oblien logs — View workspace logs

  Usage:
    oblien logs <ws-id>                          Get logs
    oblien logs get <ws-id> [--source <src>]      Get logs with source filter
    oblien logs boot <ws-id>                     Stream boot logs
    oblien logs cmd <ws-id>                      Stream command logs
    oblien logs files <ws-id>                    List log files
    oblien logs file <ws-id> <name>              Get specific log file
    oblien logs clear <ws-id>                    Clear logs

  Flags:
    --source <src>       Filter by source
    --tail <n>           Number of tail lines for streaming

  Examples:
    oblien logs ws_abc123
    oblien logs boot ws_abc123
    oblien logs boot ws_abc123 --tail 50
`);
    }
}
function requireId(id, cmd) {
    if (!id || id.startsWith('-')) {
        console.error(`Usage: oblien logs ${cmd} <workspace-id>`);
        process.exit(1);
    }
    return id;
}
function parseFlag(args, flag) {
    const idx = args.indexOf(flag);
    return idx >= 0 && idx + 1 < args.length ? args[idx + 1] : undefined;
}
async function logsGet(id, args, flags) {
    const wsId = requireId(id, 'get');
    const client = buildClient(flags);
    const params = {};
    const source = parseFlag(args, '--source');
    const tail = parseFlag(args, '--tail');
    if (source)
        params.source = source;
    if (tail)
        params.tail_lines = Number(tail);
    const res = await client.workspaces.logs.get(wsId, params);
    printResult(res, flags.json);
}
async function logsStream(type, id, args, flags) {
    const wsId = requireId(id, type);
    const client = buildClient(flags);
    const tail = parseFlag(args, '--tail');
    const params = tail ? { tail_lines: Number(tail) } : undefined;
    const stream = type === 'boot'
        ? client.workspaces.logs.streamBoot(wsId, params)
        : client.workspaces.logs.streamCmd(wsId, params);
    for await (const event of stream) {
        if (flags.json) {
            printResult(event, true);
        }
        else {
            const ts = event.timestamp ? `[${event.timestamp}] ` : '';
            process.stdout.write(`${ts}${event.message}\n`);
        }
    }
}
async function logsFiles(id, flags) {
    const wsId = requireId(id, 'files');
    const client = buildClient(flags);
    const res = await client.workspaces.logs.listFiles(wsId);
    printResult(res, flags.json);
}
async function logsFile(id, name, flags) {
    const wsId = requireId(id, 'file');
    if (!name) {
        console.error('Usage: oblien logs file <ws-id> <name>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const res = await client.workspaces.logs.getFile(wsId, name);
    printResult(res, flags.json);
}
async function logsClear(id, flags) {
    const wsId = requireId(id, 'clear');
    const client = buildClient(flags);
    await client.workspaces.logs.clear(wsId);
    console.log(`✓ Cleared logs for ${wsId}`);
}
