// oblien/dist/cli/commands/metadata.js
import { buildClient } from '../auth.js';
import { printResult } from '../output.js';
export async function metadataCmd(args, flags) {
    const sub = args[0];
    switch (sub) {
        case 'get': return mdGet(args[1], flags);
        case 'update':
        case 'set': return mdUpdate(args[1], args.slice(2), flags);
        case 'patch': return mdPatch(args[1], args.slice(2), flags);
        default:
            console.log(`
  oblien metadata — Manage workspace metadata (key-value tags)

  Usage:
    oblien metadata get <ws-id>                    Get all metadata
    oblien metadata set <ws-id> <key>=<val> ...    Set metadata (full replace)
    oblien metadata patch <ws-id> <key>=<val> ...  Merge metadata (add/update keys)

  Examples:
    oblien metadata get ws_abc123
    oblien metadata set ws_abc123 env=production team=backend
    oblien metadata patch ws_abc123 version=2.0
`);
    }
}
function requireId(id, cmd) {
    if (!id || id.startsWith('-')) {
        console.error(`Usage: oblien metadata ${cmd} <workspace-id>`);
        process.exit(1);
    }
    return id;
}
function parseKvPairs(args) {
    const result = {};
    for (const arg of args) {
        if (arg.startsWith('-'))
            continue;
        const eq = arg.indexOf('=');
        if (eq >= 0) {
            result[arg.slice(0, eq)] = arg.slice(eq + 1);
        }
    }
    return result;
}
async function mdGet(id, flags) {
    const wsId = requireId(id, 'get');
    const client = buildClient(flags);
    const res = await client.workspaces.metadata.get(wsId);
    printResult(res, flags.json);
}
async function mdUpdate(id, args, flags) {
    const wsId = requireId(id, 'set');
    const metadata = parseKvPairs(args);
    if (Object.keys(metadata).length === 0) {
        console.error('Usage: oblien metadata set <ws-id> key=value ...');
        process.exit(1);
    }
    const client = buildClient(flags);
    const res = await client.workspaces.metadata.update(wsId, { metadata });
    printResult(res, flags.json);
}
async function mdPatch(id, args, flags) {
    const wsId = requireId(id, 'patch');
    const metadata = parseKvPairs(args);
    if (Object.keys(metadata).length === 0) {
        console.error('Usage: oblien metadata patch <ws-id> key=value ...');
        process.exit(1);
    }
    const client = buildClient(flags);
    const res = await client.workspaces.metadata.patch(wsId, { metadata });
    printResult(res, flags.json);
}
