// oblien/dist/cli/commands/public-access.js
import { buildClient } from '../auth.js';
import { printResult, printSuccess, printTable } from '../output.js';
export async function publicAccessCmd(args, flags) {
    const sub = args[0];
    switch (sub) {
        case 'list':
        case 'ls': return paList(args[1], flags);
        case 'expose': return paExpose(args[1], args.slice(2), flags);
        case 'revoke': return paRevoke(args[1], args[2], flags);
        default:
            console.log(`
  oblien public-access — Expose and revoke public ports

  Usage:
    oblien public-access list <ws-id>                   List exposed ports
    oblien public-access expose <ws-id> [flags]         Expose a port
    oblien public-access revoke <ws-id> <port>          Revoke a port

  Expose flags:
    --port <number>          Port number (required)
    --protocol <proto>       Protocol (http, https, tcp)
    --label <label>          Label for the exposed port

  Examples:
    oblien public-access list ws_abc123
    oblien public-access expose ws_abc123 --port 3000 --protocol https
    oblien public-access revoke ws_abc123 3000
`);
    }
}
function requireId(id, cmd) {
    if (!id || id.startsWith('-')) {
        console.error(`Usage: oblien public-access ${cmd} <workspace-id>`);
        process.exit(1);
    }
    return id;
}
function parseFlag(args, flag) {
    const idx = args.indexOf(flag);
    return idx >= 0 && idx + 1 < args.length ? args[idx + 1] : undefined;
}
async function paList(id, flags) {
    const wsId = requireId(id, 'list');
    const client = buildClient(flags);
    const res = await client.workspaces.publicAccess.list(wsId);
    if (flags.json) {
        printResult(res, true);
    }
    else {
        const rows = (Array.isArray(res) ? res : []).map((p) => ({
            port: p.port,
            protocol: p.protocol ?? '-',
            url: p.url ?? '-',
            label: p.label ?? '-',
        }));
        if (rows.length === 0) {
            console.log('No exposed ports.');
        }
        else {
            printTable(rows, ['port', 'protocol', 'url', 'label']);
        }
    }
}
async function paExpose(id, args, flags) {
    const wsId = requireId(id, 'expose');
    const port = parseFlag(args, '--port');
    if (!port) {
        console.error('Usage: oblien public-access expose <ws-id> --port <number>');
        process.exit(1);
    }
    const params = { port: Number(port) };
    const protocol = parseFlag(args, '--protocol');
    const label = parseFlag(args, '--label');
    if (protocol)
        params.protocol = protocol;
    if (label)
        params.label = label;
    const client = buildClient(flags);
    const res = await client.workspaces.publicAccess.expose(wsId, params);
    printResult(res, flags.json);
}
async function paRevoke(id, port, flags) {
    const wsId = requireId(id, 'revoke');
    if (!port) {
        console.error('Usage: oblien public-access revoke <ws-id> <port>');
        process.exit(1);
    }
    const client = buildClient(flags);
    await client.workspaces.publicAccess.revoke(wsId, Number(port));
    printSuccess(`Revoked public access on port ${port} for ${wsId}`);
}
