// oblien/dist/cli/commands/search.js
import { buildClient } from '../auth.js';
import { printResult } from '../output.js';
export async function searchCmd(args, flags) {
    const sub = args[0];
    // oblien search files <ws-id> <query> — filename search
    // oblien search <ws-id> <query> — content search (default)
    if (sub === 'files') {
        return searchFiles(args[1], args.slice(2), flags);
    }
    if (sub === 'status') {
        return searchStatus(args[1], flags);
    }
    if (sub === 'init') {
        return searchInit(args[1], flags);
    }
    if (sub === '--help' || sub === '-h' || !sub) {
        console.log(`
  oblien search — Search in a workspace

  Usage:
    oblien search <ws-id> <query> [flags]              Content search (ripgrep)
    oblien search files <ws-id> <query> [flags]        Filename search
    oblien search status <ws-id>                       Check search engine status
    oblien search init <ws-id>                         Initialize search engine

  Content search flags:
    --dir <path>         Search directory (default: /)
    --regex              Treat query as regex
    --whole-word         Match whole words only
    --context <n>        Lines of context around matches

  Filename search flags:
    --dir <path>         Search directory (default: /)

  Examples:
    oblien search ws_abc123 "TODO"
    oblien search ws_abc123 "function.*main" --regex
    oblien search files ws_abc123 "*.ts"
`);
        return;
    }
    return searchContent(sub, args.slice(1), flags);
}
function parseFlag(args, flag) {
    const idx = args.indexOf(flag);
    return idx >= 0 && idx + 1 < args.length ? args[idx + 1] : undefined;
}
async function searchContent(wsId, args, flags) {
    const query = args.find(a => !a.startsWith('-'));
    if (!query) {
        console.error('Usage: oblien search <ws-id> <query>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const params = { query };
    const dir = parseFlag(args, '--dir');
    const context = parseFlag(args, '--context');
    if (dir)
        params.path = dir;
    if (args.includes('--regex'))
        params.regex = true;
    if (args.includes('--whole-word'))
        params.wholeWord = true;
    if (context)
        params.contextLines = Number(context);
    const res = await rt.search.content(params);
    if (flags.json) {
        printResult(res, true);
    }
    else {
        const matches = res.matches ?? [];
        if (matches.length === 0) {
            console.log('No matches found.');
            return;
        }
        for (const m of matches) {
            console.log(`${m.path}:${m.line_number ?? ''}  ${m.line ?? m.text ?? ''}`);
        }
        console.log(`\n${matches.length} match(es)`);
    }
}
async function searchFiles(wsId, args, flags) {
    if (!wsId) {
        console.error('Usage: oblien search files <ws-id> <query>');
        process.exit(1);
    }
    const query = args.find(a => !a.startsWith('-'));
    if (!query) {
        console.error('Usage: oblien search files <ws-id> <query>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const params = { query };
    const dir = parseFlag(args, '--dir');
    if (dir)
        params.path = dir;
    const res = await rt.search.files(params);
    if (flags.json) {
        printResult(res, true);
    }
    else {
        const files = res.files ?? [];
        if (files.length === 0) {
            console.log('No files found.');
            return;
        }
        for (const f of files) {
            console.log(typeof f === 'string' ? f : f.path ?? f.name ?? JSON.stringify(f));
        }
    }
}
async function searchStatus(wsId, flags) {
    if (!wsId) {
        console.error('Usage: oblien search status <ws-id>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const res = await rt.search.status();
    printResult(res, flags.json);
}
async function searchInit(wsId, flags) {
    if (!wsId) {
        console.error('Usage: oblien search init <ws-id>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const res = await rt.search.init();
    printResult(res, flags.json);
}
