// oblien/dist/cli/commands/skills.js
import { existsSync, mkdirSync, writeFileSync, rmSync } from 'node:fs';
import { join } from 'node:path';
import { printSuccess, printError, printTable } from '../output.js';
const SKILLS_BASE_URL = 'https://oblien.com/skills';
const MANIFEST_URL = `${SKILLS_BASE_URL}/manifest.json`;
const ALL_SKILLS = [
    'oblien-runtime',
    'oblien-api',
    'oblien-networking',
    'oblien-platform',
    'oblien-files',
    'oblien-search',
    'oblien-exec',
    'oblien-terminal',
    'oblien-watcher',
];
const AGENTS = {
    claude: root => join(root, '.claude', 'skills'),
    cursor: root => join(root, '.cursor', 'skills'),
    openclaw: root => join(root, 'skills'),
    goose: root => join(root, '.goose', 'skills'),
    jetbrains: root => join(root, '.jb-ai', 'skills'),
};
function detectAgent(root) {
    // Check for agent directories in order of priority
    if (existsSync(join(root, '.claude')))
        return { name: 'claude', dir: AGENTS.claude(root) };
    if (existsSync(join(root, '.cursor')))
        return { name: 'cursor', dir: AGENTS.cursor(root) };
    if (existsSync(join(root, '.goose')))
        return { name: 'goose', dir: AGENTS.goose(root) };
    if (existsSync(join(root, '.jb-ai')))
        return { name: 'jetbrains', dir: AGENTS.jetbrains(root) };
    // Default to Claude Code (most common)
    return { name: 'claude', dir: AGENTS.claude(root) };
}
function resolveSkillsDir(args) {
    let agent;
    let customPath;
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--agent' && i + 1 < args.length)
            agent = args[++i];
        else if (args[i] === '--path' && i + 1 < args.length)
            customPath = args[++i];
    }
    if (customPath)
        return customPath;
    const root = process.cwd();
    if (agent) {
        const resolver = AGENTS[agent];
        if (!resolver) {
            console.error(`Unknown agent: ${agent}\nSupported: ${Object.keys(AGENTS).join(', ')}`);
            process.exit(1);
        }
        return resolver(root);
    }
    const detected = detectAgent(root);
    return detected.dir;
}
// ── Commands ────────────────────────────────────────────────────────
export async function skillsCmd(args, _flags) {
    const sub = args[0];
    switch (sub) {
        case 'install':
            return installSkills(args.slice(1));
        case 'list':
            return listSkills();
        case 'remove':
            return removeSkill(args.slice(1));
        default:
            console.log(`
  oblien skills — Manage agent skills

  Usage:
    oblien skills install <name...>           Install skill(s)
    oblien skills install --all               Install all skills
    oblien skills list                        List available skills
    oblien skills remove <name>               Remove an installed skill

  Flags:
    --agent <name>     Target agent (claude, cursor, openclaw, goose, jetbrains)
    --path <dir>       Custom install directory
    --all              Install all 9 skills

  Examples:
    oblien skills install oblien-runtime
    oblien skills install oblien-runtime oblien-api --agent claude
    oblien skills install --all --agent cursor
    oblien skills remove oblien-runtime
`);
    }
}
async function installSkills(args) {
    const isAll = args.includes('--all');
    const names = isAll
        ? ALL_SKILLS
        : args.filter(a => !a.startsWith('--') && !(args[args.indexOf(a) - 1]?.startsWith('--')));
    if (names.length === 0) {
        console.error('Usage: oblien skills install <name...> [--agent <agent>] [--all]');
        process.exit(1);
    }
    const skillsDir = resolveSkillsDir(args);
    for (const name of names) {
        if (!ALL_SKILLS.includes(name) && !isAll) {
            printError(`Unknown skill: ${name}`);
            console.error(`Available: ${ALL_SKILLS.join(', ')}`);
            process.exit(1);
        }
        const url = `${SKILLS_BASE_URL}/${name}/SKILL.md`;
        const dest = join(skillsDir, name);
        try {
            const res = await fetch(url);
            if (!res.ok)
                throw new Error(`HTTP ${res.status}`);
            const content = await res.text();
            mkdirSync(dest, { recursive: true });
            writeFileSync(join(dest, 'SKILL.md'), content, 'utf-8');
            printSuccess(`${name} → ${dest}/SKILL.md`);
        }
        catch (err) {
            printError(`Failed to install ${name}: ${err instanceof Error ? err.message : err}`);
        }
    }
}
async function listSkills() {
    try {
        const res = await fetch(MANIFEST_URL);
        if (!res.ok)
            throw new Error(`HTTP ${res.status}`);
        const manifest = await res.json();
        printTable(manifest.skills.map(s => ({ name: s.name, description: s.description })), ['name', 'description']);
    }
    catch {
        // Fallback — just list the known skills
        console.log('Available skills:\n');
        for (const name of ALL_SKILLS) {
            console.log(`  ${name}`);
        }
    }
}
async function removeSkill(args) {
    const name = args.find(a => !a.startsWith('--'));
    if (!name) {
        console.error('Usage: oblien skills remove <name> [--agent <agent>]');
        process.exit(1);
    }
    const skillsDir = resolveSkillsDir(args);
    const dest = join(skillsDir, name);
    if (!existsSync(dest)) {
        printError(`Skill not found: ${dest}`);
        process.exit(1);
    }
    rmSync(dest, { recursive: true });
    printSuccess(`Removed ${name} from ${skillsDir}`);
}
