// oblien/dist/cli/commands/snapshots.js
import { buildClient } from '../auth.js';
import { printResult, printSuccess } from '../output.js';
export async function snapshotsCmd(args, flags) {
    const sub = args[0];
    switch (sub) {
        case 'create': return snapCreate(args[1], args.slice(2), flags);
        case 'restore': return snapRestore(args[1], flags);
        case 'archives':
        case 'list-archives': return archiveList(args[1], flags);
        case 'create-archive': return archiveCreate(args[1], args.slice(2), flags);
        case 'get-archive': return archiveGet(args[1], args[2], flags);
        case 'delete-archive': return archiveDelete(args[1], args[2], args.slice(3), flags);
        case 'delete-all-archives': return archiveDeleteAll(args[1], args.slice(2), flags);
        default:
            console.log(`
  oblien snapshots — Manage snapshots and archives

  Usage:
    oblien snapshots create <ws-id> [--after resume|stop]     Create snapshot
    oblien snapshots restore <ws-id>                          Restore from snapshot
    oblien snapshots archives <ws-id>                         List archives
    oblien snapshots create-archive <ws-id> [--version <v>]   Create archive
    oblien snapshots get-archive <ws-id> <version>            Get archive info
    oblien snapshots delete-archive <ws-id> <version>         Delete archive
    oblien snapshots delete-all-archives <ws-id>              Delete all archives

  Examples:
    oblien snapshots create ws_abc123
    oblien snapshots create ws_abc123 --after stop
    oblien snapshots archives ws_abc123
    oblien snapshots create-archive ws_abc123 --version v1
`);
    }
}
function requireId(id, cmd) {
    if (!id || id.startsWith('-')) {
        console.error(`Usage: oblien snapshots ${cmd} <workspace-id>`);
        process.exit(1);
    }
    return id;
}
function parseFlag(args, flag) {
    const idx = args.indexOf(flag);
    return idx >= 0 && idx + 1 < args.length ? args[idx + 1] : undefined;
}
async function snapCreate(id, args, flags) {
    const wsId = requireId(id, 'create');
    const client = buildClient(flags);
    const after = parseFlag(args, '--after');
    const res = await client.workspaces.snapshots.create(wsId, after ? { after } : {});
    printResult(res, flags.json);
}
async function snapRestore(id, flags) {
    const wsId = requireId(id, 'restore');
    const client = buildClient(flags);
    await client.workspaces.snapshots.restore(wsId);
    printSuccess(`Restored snapshot for ${wsId}`);
}
async function archiveList(id, flags) {
    const wsId = requireId(id, 'archives');
    const client = buildClient(flags);
    const res = await client.workspaces.snapshots.listArchives(wsId);
    printResult(res, flags.json);
}
async function archiveCreate(id, args, flags) {
    const wsId = requireId(id, 'create-archive');
    const client = buildClient(flags);
    const version = parseFlag(args, '--version');
    const res = await client.workspaces.snapshots.createArchive(wsId, version ? { version } : {});
    printResult(res, flags.json);
}
async function archiveGet(id, version, flags) {
    const wsId = requireId(id, 'get-archive');
    if (!version) {
        console.error('Usage: oblien snapshots get-archive <ws-id> <version>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const res = await client.workspaces.snapshots.getArchive(wsId, version);
    printResult(res, flags.json);
}
async function archiveDelete(id, version, args, flags) {
    const wsId = requireId(id, 'delete-archive');
    if (!version) {
        console.error('Usage: oblien snapshots delete-archive <ws-id> <version>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const deleteFile = args.includes('--delete-file');
    await client.workspaces.snapshots.deleteArchive(wsId, version, deleteFile ? { delete_file: true } : undefined);
    printSuccess(`Deleted archive ${version} for ${wsId}`);
}
async function archiveDeleteAll(id, args, flags) {
    const wsId = requireId(id, 'delete-all-archives');
    const client = buildClient(flags);
    const deleteFiles = args.includes('--delete-files');
    const res = await client.workspaces.snapshots.deleteAllArchives(wsId, deleteFiles ? { delete_files: true } : undefined);
    printResult(res, flags.json);
}
