// oblien/dist/cli/commands/ssh.js
import { buildClient } from '../auth.js';
import { printResult, printSuccess } from '../output.js';
export async function sshCmd(args, flags) {
    const sub = args[0];
    switch (sub) {
        case 'status': return sshStatus(args[1], flags);
        case 'enable': return sshEnable(args[1], flags);
        case 'disable': return sshDisable(args[1], flags);
        case 'set-password': return sshSetPassword(args[1], args.slice(2), flags);
        case 'set-key': return sshSetKey(args[1], args.slice(2), flags);
        default:
            console.log(`
  oblien ssh — Manage SSH access

  Usage:
    oblien ssh status <ws-id>                       Check SSH status
    oblien ssh enable <ws-id>                       Enable SSH
    oblien ssh disable <ws-id>                      Disable SSH
    oblien ssh set-password <ws-id> --password <p>  Set SSH password
    oblien ssh set-key <ws-id> --key <public-key>   Set SSH public key

  Examples:
    oblien ssh status ws_abc123
    oblien ssh enable ws_abc123
    oblien ssh set-password ws_abc123 --password mysecret
    oblien ssh set-key ws_abc123 --key "ssh-ed25519 AAAA..."
`);
    }
}
function requireId(id, cmd) {
    if (!id || id.startsWith('-')) {
        console.error(`Usage: oblien ssh ${cmd} <workspace-id>`);
        process.exit(1);
    }
    return id;
}
function parseFlag(args, flag) {
    const idx = args.indexOf(flag);
    return idx >= 0 && idx + 1 < args.length ? args[idx + 1] : undefined;
}
async function sshStatus(id, flags) {
    const wsId = requireId(id, 'status');
    const client = buildClient(flags);
    const res = await client.workspaces.ssh.status(wsId);
    printResult(res, flags.json);
}
async function sshEnable(id, flags) {
    const wsId = requireId(id, 'enable');
    const client = buildClient(flags);
    await client.workspaces.ssh.enable(wsId);
    printSuccess(`SSH enabled for ${wsId}`);
}
async function sshDisable(id, flags) {
    const wsId = requireId(id, 'disable');
    const client = buildClient(flags);
    await client.workspaces.ssh.disable(wsId);
    printSuccess(`SSH disabled for ${wsId}`);
}
async function sshSetPassword(id, args, flags) {
    const wsId = requireId(id, 'set-password');
    const password = parseFlag(args, '--password');
    if (!password) {
        console.error('Usage: oblien ssh set-password <ws-id> --password <password>');
        process.exit(1);
    }
    const client = buildClient(flags);
    await client.workspaces.ssh.setPassword(wsId, { password });
    printSuccess(`SSH password set for ${wsId}`);
}
async function sshSetKey(id, args, flags) {
    const wsId = requireId(id, 'set-key');
    const key = parseFlag(args, '--key');
    if (!key) {
        console.error('Usage: oblien ssh set-key <ws-id> --key <public-key>');
        process.exit(1);
    }
    const client = buildClient(flags);
    await client.workspaces.ssh.setKey(wsId, { public_key: key });
    printSuccess(`SSH public key set for ${wsId}`);
}
