// oblien/dist/cli/commands/terminal.js
import { buildClient } from '../auth.js';
import { printResult, printSuccess, printTable } from '../output.js';
export async function terminalCmd(args, flags) {
    const sub = args[0];
    switch (sub) {
        case 'create': return termCreate(args[1], args.slice(2), flags);
        case 'list':
        case 'ls': return termList(args[1], flags);
        case 'close': return termClose(args[1], args[2], flags);
        case 'scrollback': return termScrollback(args[1], args[2], args.slice(3), flags);
        default:
            console.log(`
  oblien terminal — Manage runtime terminal sessions

  Usage:
    oblien terminal create <ws-id> [--shell <sh>] [--cols <n>] [--rows <n>]
    oblien terminal list <ws-id>              List sessions
    oblien terminal close <ws-id> <session>   Close session
    oblien terminal scrollback <ws-id> <session> [--bytes <n>]

  Examples:
    oblien terminal create ws_abc123
    oblien terminal create ws_abc123 --shell /bin/bash --cols 120 --rows 40
    oblien terminal list ws_abc123
    oblien terminal close ws_abc123 sess_def456
    oblien terminal scrollback ws_abc123 sess_def456 --bytes 4096
`);
    }
}
function requireWsId(id, cmd) {
    if (!id || id.startsWith('-')) {
        console.error(`Usage: oblien terminal ${cmd} <workspace-id>`);
        process.exit(1);
    }
    return id;
}
function parseFlag(args, flag) {
    const idx = args.indexOf(flag);
    return idx >= 0 && idx + 1 < args.length ? args[idx + 1] : undefined;
}
async function termCreate(wsIdRaw, args, flags) {
    const wsId = requireWsId(wsIdRaw, 'create');
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const params = {};
    const shell = parseFlag(args, '--shell');
    const cols = parseFlag(args, '--cols');
    const rows = parseFlag(args, '--rows');
    if (shell)
        params.shell = shell;
    if (cols)
        params.cols = Number(cols);
    if (rows)
        params.rows = Number(rows);
    const res = await rt.terminal.create(params);
    printResult(res, flags.json);
}
async function termList(wsIdRaw, flags) {
    const wsId = requireWsId(wsIdRaw, 'list');
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const res = await rt.terminal.list();
    if (flags.json) {
        printResult(res, true);
    }
    else {
        const items = Array.isArray(res) ? res : [];
        if (items.length === 0) {
            console.log('No terminal sessions.');
        }
        else {
            const rows = items.map((s) => ({
                id: s.id ?? s.terminal_id ?? '-',
                shell: s.shell ?? '-',
                status: s.status ?? '-',
            }));
            printTable(rows, ['id', 'shell', 'status']);
        }
    }
}
async function termClose(wsIdRaw, sessionId, flags) {
    const wsId = requireWsId(wsIdRaw, 'close');
    if (!sessionId) {
        console.error('Usage: oblien terminal close <ws-id> <session-id>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    await rt.terminal.close(sessionId);
    printSuccess(`Closed terminal ${sessionId}`);
}
async function termScrollback(wsIdRaw, sessionId, args, flags) {
    const wsId = requireWsId(wsIdRaw, 'scrollback');
    if (!sessionId) {
        console.error('Usage: oblien terminal scrollback <ws-id> <session-id>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const bytes = parseFlag(args, '--bytes');
    const res = await rt.terminal.scrollback(sessionId, bytes ? { bytes: Number(bytes) } : undefined);
    if (flags.json) {
        printResult(res, true);
    }
    else {
        process.stdout.write(res.data ?? res.content ?? JSON.stringify(res));
    }
}
