// oblien/dist/cli/commands/usage.js
import { buildClient } from '../auth.js';
import { printResult, printSuccess } from '../output.js';
export async function usageCmd(args, flags) {
    const sub = args[0];
    switch (sub) {
        case 'get': return uGet(args[1], args.slice(2), flags);
        case 'totals': return uTotals(args[1], flags);
        case 'credits':
        case 'credits-chart': return uCreditsChart(args[1], flags);
        case 'global': return uGlobal(flags);
        case 'activity': return uActivity(flags);
        case 'wipe': return uWipe(args[1], flags);
        case 'start-tracking': return uStartTracking(args[1], flags);
        case 'stop-tracking': return uStopTracking(args[1], flags);
        default:
            console.log(`
  oblien usage — View usage and credits

  Usage:
    oblien usage get <ws-id> [--period <p>]    Get workspace usage
    oblien usage totals <ws-id>                Get usage totals
    oblien usage credits <ws-id>               Credits chart data
    oblien usage global                        Global usage across all workspaces
    oblien usage activity                      Account activity
    oblien usage wipe <ws-id>                  Reset usage data
    oblien usage start-tracking <ws-id>        Start tracking usage
    oblien usage stop-tracking <ws-id>         Stop tracking usage

  Examples:
    oblien usage get ws_abc123 --period 7d
    oblien usage totals ws_abc123
    oblien usage global
    oblien usage activity --json
`);
    }
}
function requireId(id, cmd) {
    if (!id || id.startsWith('-')) {
        console.error(`Usage: oblien usage ${cmd} <workspace-id>`);
        process.exit(1);
    }
    return id;
}
function parseFlag(args, flag) {
    const idx = args.indexOf(flag);
    return idx >= 0 && idx + 1 < args.length ? args[idx + 1] : undefined;
}
async function uGet(id, args, flags) {
    const wsId = requireId(id, 'get');
    const client = buildClient(flags);
    const params = {};
    const period = parseFlag(args, '--period');
    if (period)
        params.period = period;
    const res = await client.workspaces.usage.get(wsId, params);
    printResult(res, flags.json);
}
async function uTotals(id, flags) {
    const wsId = requireId(id, 'totals');
    const client = buildClient(flags);
    const res = await client.workspaces.usage.totals(wsId);
    printResult(res, flags.json);
}
async function uCreditsChart(id, flags) {
    const wsId = requireId(id, 'credits');
    const client = buildClient(flags);
    const res = await client.workspaces.usage.creditsChart(wsId);
    printResult(res, flags.json);
}
async function uGlobal(flags) {
    const client = buildClient(flags);
    const res = await client.workspaces.usage.global();
    printResult(res, flags.json);
}
async function uActivity(flags) {
    const client = buildClient(flags);
    const res = await client.workspaces.usage.activity();
    printResult(res, flags.json);
}
async function uWipe(id, flags) {
    const wsId = requireId(id, 'wipe');
    const client = buildClient(flags);
    await client.workspaces.usage.wipe(wsId);
    printSuccess(`Wiped usage data for ${wsId}`);
}
async function uStartTracking(id, flags) {
    const wsId = requireId(id, 'start-tracking');
    const client = buildClient(flags);
    await client.workspaces.usage.startTracking(wsId);
    printSuccess(`Started usage tracking for ${wsId}`);
}
async function uStopTracking(id, flags) {
    const wsId = requireId(id, 'stop-tracking');
    const client = buildClient(flags);
    await client.workspaces.usage.stopTracking(wsId);
    printSuccess(`Stopped usage tracking for ${wsId}`);
}
