// oblien/dist/cli/commands/watcher.js
import { buildClient } from '../auth.js';
import { printResult, printSuccess, printTable } from '../output.js';
export async function watcherCmd(args, flags) {
    const sub = args[0];
    switch (sub) {
        case 'create': return wCreate(args[1], args.slice(2), flags);
        case 'list':
        case 'ls': return wList(args[1], flags);
        case 'get': return wGet(args[1], args[2], flags);
        case 'delete':
        case 'rm': return wDelete(args[1], args[2], flags);
        default:
            console.log(`
  oblien watcher — Manage file system watchers

  Usage:
    oblien watcher create <ws-id> --path <p> [--recursive] [--events <e>]
    oblien watcher list <ws-id>                 List active watchers
    oblien watcher get <ws-id> <watcher-id>     Get watcher info
    oblien watcher delete <ws-id> <watcher-id>  Remove watcher

  Create flags:
    --path <path>            Path to watch (required)
    --recursive              Watch subdirectories
    --events <events>        Comma-separated: create,modify,delete

  Examples:
    oblien watcher create ws_abc123 --path /app/src --recursive
    oblien watcher list ws_abc123
    oblien watcher delete ws_abc123 w_def456
`);
    }
}
function requireWsId(id, cmd) {
    if (!id || id.startsWith('-')) {
        console.error(`Usage: oblien watcher ${cmd} <workspace-id>`);
        process.exit(1);
    }
    return id;
}
function parseFlag(args, flag) {
    const idx = args.indexOf(flag);
    return idx >= 0 && idx + 1 < args.length ? args[idx + 1] : undefined;
}
async function wCreate(wsIdRaw, args, flags) {
    const wsId = requireWsId(wsIdRaw, 'create');
    const path = parseFlag(args, '--path');
    if (!path) {
        console.error('Usage: oblien watcher create <ws-id> --path <path>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const params = { path };
    if (args.includes('--recursive'))
        params.recursive = true;
    const events = parseFlag(args, '--events');
    if (events)
        params.events = events.split(',');
    const res = await rt.watcher.create(params);
    printResult(res, flags.json);
}
async function wList(wsIdRaw, flags) {
    const wsId = requireWsId(wsIdRaw, 'list');
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const res = await rt.watcher.list();
    if (flags.json) {
        printResult(res, true);
    }
    else {
        const items = Array.isArray(res) ? res : [];
        if (items.length === 0) {
            console.log('No active watchers.');
        }
        else {
            const rows = items.map((w) => ({
                id: w.id ?? w.watcher_id ?? '-',
                path: w.path ?? '-',
                recursive: w.recursive ? 'yes' : 'no',
            }));
            printTable(rows, ['id', 'path', 'recursive']);
        }
    }
}
async function wGet(wsIdRaw, watcherId, flags) {
    const wsId = requireWsId(wsIdRaw, 'get');
    if (!watcherId) {
        console.error('Usage: oblien watcher get <ws-id> <watcher-id>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const res = await rt.watcher.get(watcherId);
    printResult(res, flags.json);
}
async function wDelete(wsIdRaw, watcherId, flags) {
    const wsId = requireWsId(wsIdRaw, 'delete');
    if (!watcherId) {
        console.error('Usage: oblien watcher delete <ws-id> <watcher-id>');
        process.exit(1);
    }
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    await rt.watcher.delete(watcherId);
    printSuccess(`Deleted watcher ${watcherId}`);
}
