// oblien/dist/cli/commands/workspaces.js
import { buildClient } from '../auth.js';
import { printResult, printSuccess, printTable } from '../output.js';
export async function workspacesCmd(args, flags) {
    const sub = args[0];
    switch (sub) {
        case 'create': return wsCreate(args.slice(1), flags);
        case 'list':
        case 'ls': return wsList(args.slice(1), flags);
        case 'get': return wsGet(args[1], flags);
        case 'update': return wsUpdate(args[1], args.slice(2), flags);
        case 'delete':
        case 'rm': return wsDelete(args[1], flags);
        case 'start': return wsPower('start', args[1], flags);
        case 'stop': return wsPower('stop', args[1], flags);
        case 'restart': return wsPower('restart', args[1], flags);
        case 'pause': return wsPower('pause', args[1], flags);
        case 'resume': return wsPower('resume', args[1], flags);
        case 'ping': return wsPing(args[1], args.slice(2), flags);
        case 'quota': return wsQuota(flags);
        case 'estimate': return wsEstimate(args.slice(1), flags);
        case 'health': return wsHealth(args[1], flags);
        case 'details': return wsDetails(args[1], flags);
        case 'resources': return wsResources(args[1], args.slice(2), flags);
        default:
            console.log(`
  oblien ws — Manage workspaces

  Usage:
    oblien ws create [flags]         Create a workspace
    oblien ws list                   List workspaces
    oblien ws get <id>               Get workspace details
    oblien ws update <id> [flags]    Update workspace
    oblien ws delete <id>            Delete workspace
    oblien ws start <id>             Start workspace
    oblien ws stop <id>              Stop workspace
    oblien ws restart <id>           Restart workspace
    oblien ws pause <id>             Pause workspace
    oblien ws resume <id>            Resume paused workspace
    oblien ws ping <id>              Keep-alive ping (renew TTL)
    oblien ws quota                  Show resource quota
    oblien ws estimate [flags]       Estimate cost
    oblien ws health <id>            Check runtime health
    oblien ws details <id>           Get aggregated details
    oblien ws resources <id>         Get/update resources (--patch for partial)

  Create flags:
    --image <image>       Base image (e.g., node-20, python-3, ubuntu)
    --name <name>         Workspace name
    --cpus <n>            CPU cores
    --memory <mb>         Memory in MB
    --disk <mb>           Disk in MB
    --mode <mode>         permanent or temporary
    --namespace <ns>      Namespace
    --type <type>         Workspace type

  Examples:
    oblien ws create --image node-20 --name my-app --cpus 2 --memory 2048
    oblien ws list --status running --limit 10
    oblien ws start ws_abc123
    oblien ws stop ws_abc123
    oblien ws resources ws_abc123 --cpus 4 --memory 4096
`);
    }
}
// ── Helpers ─────────────────────────────────────────────────────────
function parseFlag(args, flag) {
    const idx = args.indexOf(flag);
    return idx >= 0 && idx + 1 < args.length ? args[idx + 1] : undefined;
}
function requireId(id, cmd) {
    if (!id || id.startsWith('-')) {
        console.error(`Usage: oblien ws ${cmd} <workspace-id>`);
        process.exit(1);
    }
    return id;
}
// ── Commands ────────────────────────────────────────────────────────
async function wsCreate(args, flags) {
    const client = buildClient(flags);
    const params = {};
    const image = parseFlag(args, '--image');
    const name = parseFlag(args, '--name');
    const cpus = parseFlag(args, '--cpus');
    const memory = parseFlag(args, '--memory');
    const disk = parseFlag(args, '--disk');
    const mode = parseFlag(args, '--mode');
    const namespace = parseFlag(args, '--namespace');
    const type = parseFlag(args, '--type');
    if (image)
        params.image = image;
    if (name)
        params.name = name;
    if (mode)
        params.mode = mode;
    if (namespace)
        params.namespace = namespace;
    if (type)
        params.type = type;
    if (cpus || memory || disk) {
        params.resources = {
            ...(cpus ? { cpus: Number(cpus) } : {}),
            ...(memory ? { memory_mb: Number(memory) } : {}),
            ...(disk ? { writable_size_mb: Number(disk) } : {}),
        };
    }
    const ws = await client.workspaces.create(params);
    printResult(ws, flags.json);
}
async function wsList(args, flags) {
    const client = buildClient(flags);
    const params = {};
    const status = parseFlag(args, '--status');
    const limit = parseFlag(args, '--limit');
    const page = parseFlag(args, '--page');
    if (status)
        params.status = status;
    if (limit)
        params.limit = Number(limit);
    if (page)
        params.page = Number(page);
    const res = await client.workspaces.list(params);
    if (flags.json) {
        printResult(res, true);
    }
    else {
        const rows = res.workspaces.map(w => ({
            id: w.id,
            name: w.name ?? '-',
            status: w.status,
            image: w.image,
            created: w.created_at?.slice(0, 10) ?? '-',
        }));
        printTable(rows, ['id', 'name', 'status', 'image', 'created']);
        console.log(`\n${res.total} workspace(s) total`);
    }
}
async function wsGet(id, flags) {
    const wsId = requireId(id, 'get');
    const client = buildClient(flags);
    const ws = await client.workspaces.get(wsId);
    printResult(ws, flags.json);
}
async function wsUpdate(id, args, flags) {
    const wsId = requireId(id, 'update');
    const client = buildClient(flags);
    // Collect all --key value pairs as update params
    const params = {};
    for (let i = 0; i < args.length; i++) {
        if (args[i].startsWith('--') && i + 1 < args.length) {
            const key = args[i].slice(2).replace(/-/g, '_');
            params[key] = args[++i];
        }
    }
    const ws = await client.workspaces.update(wsId, params);
    printResult(ws, flags.json);
}
async function wsDelete(id, flags) {
    const wsId = requireId(id, 'delete');
    const client = buildClient(flags);
    await client.workspaces.delete(wsId);
    printSuccess(`Deleted workspace ${wsId}`);
}
async function wsPower(action, id, flags) {
    const wsId = requireId(id, action);
    const client = buildClient(flags);
    switch (action) {
        case 'start':
            await client.workspaces.start(wsId);
            break;
        case 'stop':
            await client.workspaces.stop(wsId);
            break;
        case 'restart':
            await client.workspaces.restart(wsId);
            break;
        case 'pause':
            await client.workspaces.pause(wsId);
            break;
        case 'resume':
            await client.workspaces.resume(wsId);
            break;
    }
    printSuccess(`${action}: ${wsId}`);
}
async function wsPing(id, args, flags) {
    const wsId = requireId(id, 'ping');
    const client = buildClient(flags);
    const ttl = parseFlag(args, '--ttl');
    await client.workspaces.ping(wsId, ttl ? { ttl_seconds: Number(ttl) } : undefined);
    printSuccess(`Pinged ${wsId}${ttl ? ` (TTL: ${ttl}s)` : ''}`);
}
async function wsQuota(flags) {
    const client = buildClient(flags);
    const res = await client.workspaces.getQuota();
    printResult(res, flags.json);
}
async function wsEstimate(args, flags) {
    const client = buildClient(flags);
    const params = {};
    const cpus = parseFlag(args, '--cpus');
    const ram = parseFlag(args, '--memory');
    const disk = parseFlag(args, '--disk');
    if (cpus)
        params.cpu = Number(cpus);
    if (ram)
        params.ram_mb = Number(ram);
    if (disk)
        params.disk_mb = Number(disk);
    const res = await client.workspaces.getEstimate(params);
    printResult(res, flags.json);
}
async function wsHealth(id, flags) {
    const wsId = requireId(id, 'health');
    const client = buildClient(flags);
    const rt = await client.workspaces.runtime(wsId);
    const ok = await rt.health();
    if (flags.json) {
        printResult({ healthy: ok }, true);
    }
    else {
        console.log(ok ? '✓ Healthy' : '✗ Unhealthy');
    }
}
async function wsDetails(id, flags) {
    const wsId = requireId(id, 'details');
    const client = buildClient(flags);
    const res = await client.workspaces.getDetails(wsId);
    printResult(res, flags.json);
}
async function wsResources(id, args, flags) {
    const wsId = requireId(id, 'resources');
    const client = buildClient(flags);
    const patch = args.includes('--patch');
    // If any flags provided — update; otherwise — get
    const cpus = parseFlag(args, '--cpus');
    const memory = parseFlag(args, '--memory');
    const disk = parseFlag(args, '--disk');
    if (cpus || memory || disk) {
        const params = {};
        if (cpus)
            params.cpus = Number(cpus);
        if (memory)
            params.memory_mb = Number(memory);
        if (disk)
            params.writable_size_mb = Number(disk);
        const method = patch ? 'patch' : 'update';
        const res = await client.workspaces.resources[method](wsId, params);
        printResult(res, flags.json);
    }
    else {
        const res = await client.workspaces.resources.get(wsId);
        printResult(res, flags.json);
    }
}
