// oblien/dist/cli/commands/workloads.js
import { buildClient } from '../auth.js';
import { printResult, printSuccess, printTable } from '../output.js';
export async function workloadsCmd(args, flags) {
    const sub = args[0];
    switch (sub) {
        case 'create': return wlCreate(args[1], args.slice(2), flags);
        case 'list':
        case 'ls': return wlList(args[1], args.slice(2), flags);
        case 'get': return wlGet(args[1], args[2], flags);
        case 'update': return wlUpdate(args[1], args[2], args.slice(3), flags);
        case 'patch': return wlPatch(args[1], args[2], args.slice(3), flags);
        case 'delete':
        case 'rm': return wlDelete(args[1], args[2], flags);
        case 'delete-all': return wlDeleteAll(args[1], flags);
        case 'start': return wlAction('start', args[1], args[2], flags);
        case 'stop': return wlAction('stop', args[1], args[2], flags);
        case 'status': return wlStatus(args[1], args[2], flags);
        case 'logs': return wlLogs(args[1], args[2], args.slice(3), flags);
        case 'stats': return wlStats(args[1], args[2], args.slice(3), flags);
        default:
            console.log(`
  oblien workloads — Manage workspace workloads (containers/processes)

  Usage:
    oblien workloads create <ws-id> [flags]                 Create workload
    oblien workloads list <ws-id>                           List workloads
    oblien workloads get <ws-id> <wl-id>                    Get workload
    oblien workloads update <ws-id> <wl-id> [flags]         Full update
    oblien workloads patch <ws-id> <wl-id> [flags]          Partial update
    oblien workloads delete <ws-id> <wl-id>                 Delete workload
    oblien workloads delete-all <ws-id>                     Delete all workloads
    oblien workloads start <ws-id> <wl-id>                  Start workload
    oblien workloads stop <ws-id> <wl-id>                   Stop workload
    oblien workloads status <ws-id> <wl-id>                 Get status
    oblien workloads logs <ws-id> <wl-id> [--stream]        View/stream logs
    oblien workloads stats <ws-id> <wl-id> [--stream]       View/stream stats

  Create flags:
    --name <name>          Workload name
    --image <image>        Container image
    --cmd <command>        Command to run
    --port <port>          Exposed port
    --env <KEY=VAL,...>    Environment variables

  Examples:
    oblien workloads create ws_abc123 --name web --image nginx --port 80
    oblien workloads list ws_abc123
    oblien workloads logs ws_abc123 wl_def456 --stream
    oblien workloads stats ws_abc123 wl_def456 --stream
`);
    }
}
function requireWsId(id, cmd) {
    if (!id || id.startsWith('-')) {
        console.error(`Usage: oblien workloads ${cmd} <workspace-id> ...`);
        process.exit(1);
    }
    return id;
}
function requireWlId(id, cmd) {
    if (!id || id.startsWith('-')) {
        console.error(`Usage: oblien workloads ${cmd} <workspace-id> <workload-id>`);
        process.exit(1);
    }
    return id;
}
function parseFlag(args, flag) {
    const idx = args.indexOf(flag);
    return idx >= 0 && idx + 1 < args.length ? args[idx + 1] : undefined;
}
async function wlCreate(wsIdRaw, args, flags) {
    const wsId = requireWsId(wsIdRaw, 'create');
    const client = buildClient(flags);
    const params = {};
    const name = parseFlag(args, '--name');
    const image = parseFlag(args, '--image');
    const cmd = parseFlag(args, '--cmd');
    const port = parseFlag(args, '--port');
    const env = parseFlag(args, '--env');
    if (name)
        params.name = name;
    if (image)
        params.image = image;
    if (cmd)
        params.command = cmd;
    if (port)
        params.port = Number(port);
    if (env) {
        params.env = Object.fromEntries(env.split(',').map(pair => {
            const eq = pair.indexOf('=');
            return eq >= 0 ? [pair.slice(0, eq), pair.slice(eq + 1)] : [pair, ''];
        }));
    }
    const res = await client.workspaces.workloads.create(wsId, params);
    printResult(res, flags.json);
}
async function wlList(wsIdRaw, args, flags) {
    const wsId = requireWsId(wsIdRaw, 'list');
    const client = buildClient(flags);
    const params = {};
    const limit = parseFlag(args, '--limit');
    if (limit)
        params.limit = Number(limit);
    const res = await client.workspaces.workloads.list(wsId, params);
    if (flags.json) {
        printResult(res, true);
    }
    else {
        const rows = (Array.isArray(res) ? res : []).map((w) => ({
            id: w.id,
            name: w.name ?? '-',
            status: w.status ?? '-',
            image: w.image ?? '-',
        }));
        if (rows.length === 0) {
            console.log('No workloads.');
        }
        else {
            printTable(rows, ['id', 'name', 'status', 'image']);
        }
    }
}
async function wlGet(wsIdRaw, wlIdRaw, flags) {
    const wsId = requireWsId(wsIdRaw, 'get');
    const wlId = requireWlId(wlIdRaw, 'get');
    const client = buildClient(flags);
    const res = await client.workspaces.workloads.get(wsId, wlId);
    printResult(res, flags.json);
}
async function wlUpdate(wsIdRaw, wlIdRaw, args, flags) {
    const wsId = requireWsId(wsIdRaw, 'update');
    const wlId = requireWlId(wlIdRaw, 'update');
    const client = buildClient(flags);
    const params = {};
    for (let i = 0; i < args.length; i++) {
        if (args[i].startsWith('--') && i + 1 < args.length) {
            const key = args[i].slice(2).replace(/-/g, '_');
            params[key] = args[++i];
        }
    }
    const res = await client.workspaces.workloads.update(wsId, wlId, params);
    printResult(res, flags.json);
}
async function wlPatch(wsIdRaw, wlIdRaw, args, flags) {
    const wsId = requireWsId(wsIdRaw, 'patch');
    const wlId = requireWlId(wlIdRaw, 'patch');
    const client = buildClient(flags);
    const params = {};
    for (let i = 0; i < args.length; i++) {
        if (args[i].startsWith('--') && i + 1 < args.length) {
            const key = args[i].slice(2).replace(/-/g, '_');
            params[key] = args[++i];
        }
    }
    const res = await client.workspaces.workloads.patch(wsId, wlId, params);
    printResult(res, flags.json);
}
async function wlDelete(wsIdRaw, wlIdRaw, flags) {
    const wsId = requireWsId(wsIdRaw, 'delete');
    const wlId = requireWlId(wlIdRaw, 'delete');
    const client = buildClient(flags);
    await client.workspaces.workloads.delete(wsId, wlId);
    printSuccess(`Deleted workload ${wlId}`);
}
async function wlDeleteAll(wsIdRaw, flags) {
    const wsId = requireWsId(wsIdRaw, 'delete-all');
    const client = buildClient(flags);
    const res = await client.workspaces.workloads.deleteAll(wsId);
    printResult(res, flags.json);
}
async function wlAction(action, wsIdRaw, wlIdRaw, flags) {
    const wsId = requireWsId(wsIdRaw, action);
    const wlId = requireWlId(wlIdRaw, action);
    const client = buildClient(flags);
    if (action === 'start') {
        await client.workspaces.workloads.start(wsId, wlId);
    }
    else {
        await client.workspaces.workloads.stop(wsId, wlId);
    }
    printSuccess(`${action}: workload ${wlId}`);
}
async function wlStatus(wsIdRaw, wlIdRaw, flags) {
    const wsId = requireWsId(wsIdRaw, 'status');
    const wlId = requireWlId(wlIdRaw, 'status');
    const client = buildClient(flags);
    const res = await client.workspaces.workloads.status(wsId, wlId);
    printResult(res, flags.json);
}
async function wlLogs(wsIdRaw, wlIdRaw, args, flags) {
    const wsId = requireWsId(wsIdRaw, 'logs');
    const wlId = requireWlId(wlIdRaw, 'logs');
    const client = buildClient(flags);
    if (args.includes('--stream')) {
        for await (const event of client.workspaces.workloads.logsStream(wsId, wlId)) {
            if (flags.json) {
                printResult(event, true);
            }
            else {
                const ts = event.timestamp ? `[${event.timestamp}] ` : '';
                process.stdout.write(`${ts}${event.message ?? event.data ?? JSON.stringify(event)}\n`);
            }
        }
    }
    else {
        const res = await client.workspaces.workloads.logs(wsId, wlId);
        printResult(res, flags.json);
    }
}
async function wlStats(wsIdRaw, wlIdRaw, args, flags) {
    const wsId = requireWsId(wsIdRaw, 'stats');
    const wlId = requireWlId(wlIdRaw, 'stats');
    const client = buildClient(flags);
    if (args.includes('--stream')) {
        if (wlId === '--all') {
            // oblien workloads stats ws_abc --all --stream → allStatsStream
            for await (const event of client.workspaces.workloads.allStatsStream(wsId)) {
                printResult(event, flags.json || true);
            }
        }
        else {
            for await (const event of client.workspaces.workloads.statsStream(wsId, wlId)) {
                printResult(event, flags.json || true);
            }
        }
    }
    else {
        const res = await client.workspaces.workloads.stats(wsId, wlId);
        printResult(res, flags.json);
    }
}
