// oblien/dist/resources/api-access.js
/** Internal API server management — enable, disable, tokens. */
export class ApiAccessResource {
    _http;
    constructor(_http) {
        this._http = _http;
    }
    /** Get server status. */
    async status(workspaceId) {
        return this._http.request({ method: 'GET', path: `/workspace/${workspaceId}/internal-api-access` });
    }
    /** Enable the internal API server. Returns a Gateway JWT. */
    async enable(workspaceId) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/internal-api-access/enable` });
    }
    /** Disable the internal API server. */
    async disable(workspaceId) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/internal-api-access/disable` });
    }
    /** Rotate the gateway token. */
    async rotateToken(workspaceId) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/internal-api-access/token` });
    }
    /** Get the current gateway JWT (read-only, no side effects). */
    async getToken(workspaceId) {
        return this._http.request({ method: 'GET', path: `/workspace/${workspaceId}/internal-api-access/token` });
    }
    /** Get the raw connection token + private IP (for direct access). */
    async rawToken(workspaceId) {
        return this._http.request({ method: 'GET', path: `/workspace/${workspaceId}/internal-api-access/token/raw` });
    }
}
