// oblien/dist/resources/lifecycle.js
/** Lifecycle management — permanent/temporary mode, TTL, ping, destroy. */
export class LifecycleResource {
    _http;
    constructor(_http) {
        this._http = _http;
    }
    /** Get lifecycle status. */
    async get(workspaceId) {
        return this._http.request({ method: 'GET', path: `/workspace/${workspaceId}/lifecycle` });
    }
    /** Destroy lifecycle state. */
    async destroy(workspaceId) {
        return this._http.request({ method: 'DELETE', path: `/workspace/${workspaceId}/lifecycle` });
    }
    /** Make workspace permanent. */
    async makePermanent(workspaceId) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/lifecycle/permanent` });
    }
    /** Make workspace temporary with TTL. */
    async makeTemporary(workspaceId, params = {}) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/lifecycle/temporary`, body: params });
    }
    /** Update TTL settings. */
    async updateTtl(workspaceId, params) {
        return this._http.request({ method: 'PUT', path: `/workspace/${workspaceId}/lifecycle/ttl`, body: params });
    }
    /** Keep-alive ping — renews TTL. */
    async ping(workspaceId, params) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/ping`, body: params });
    }
}
