// oblien/dist/resources/logs.js
/** Workspace logs — boot logs, command logs, log files, streaming. */
export class LogsResource {
    _http;
    constructor(_http) {
        this._http = _http;
    }
    /** Get logs. */
    async get(workspaceId, params = {}) {
        return this._http.request({
            method: 'GET',
            path: `/workspace/${workspaceId}/logs`,
            query: params,
        });
    }
    /** Clear logs. */
    async clear(workspaceId) {
        return this._http.request({ method: 'DELETE', path: `/workspace/${workspaceId}/logs` });
    }
    /** List available log files. */
    async listFiles(workspaceId) {
        const res = await this._http.request({
            method: 'GET',
            path: `/workspace/${workspaceId}/logs/list`,
        });
        return res.files ?? [];
    }
    /** Get a specific log file by name. */
    async getFile(workspaceId, name) {
        return this._http.request({
            method: 'GET',
            path: `/workspace/${workspaceId}/logs/file/${name}`,
        });
    }
    /** Stream boot logs (SSE). */
    async *streamBoot(workspaceId, params) {
        yield* this._http.requestSSE({
            method: 'GET',
            path: `/workspace/${workspaceId}/logs/stream/boot`,
            query: params?.tail_lines ? { tail_lines: params.tail_lines } : undefined,
        });
    }
    /** Stream command logs (SSE). */
    async *streamCmd(workspaceId, params) {
        yield* this._http.requestSSE({
            method: 'GET',
            path: `/workspace/${workspaceId}/logs/stream/cmd`,
            query: params?.tail_lines ? { tail_lines: params.tail_lines } : undefined,
        });
    }
}
