// oblien/dist/resources/snapshots.js
/** Snapshots and archives. */
export class SnapshotsResource {
    _http;
    constructor(_http) {
        this._http = _http;
    }
    /** Create a snapshot. */
    async create(workspaceId, params = {}) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/snapshot`, body: params });
    }
    /** Restore from the latest snapshot. */
    async restore(workspaceId) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/restore` });
    }
    /** Create an archive. */
    async createArchive(workspaceId, params = {}) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/archives`, body: params });
    }
    /** List archives. */
    async listArchives(workspaceId) {
        const res = await this._http.request({ method: 'GET', path: `/workspace/${workspaceId}/archives` });
        return res.archives ?? [];
    }
    /** Get a specific archive. */
    async getArchive(workspaceId, version) {
        return this._http.request({ method: 'GET', path: `/workspace/${workspaceId}/archives/${version}` });
    }
    /** Delete a specific archive. */
    async deleteArchive(workspaceId, version, params) {
        return this._http.request({
            method: 'DELETE',
            path: `/workspace/${workspaceId}/archives/${version}`,
            query: params?.delete_file ? { delete_file: 'true' } : undefined,
        });
    }
    /** Delete all archives. */
    async deleteAllArchives(workspaceId, params) {
        return this._http.request({
            method: 'DELETE',
            path: `/workspace/${workspaceId}/archives`,
            query: params?.delete_files ? { delete_files: 'true' } : undefined,
        });
    }
}
