// oblien/dist/resources/ssh.js
/** SSH access management. */
export class SshResource {
    _http;
    constructor(_http) {
        this._http = _http;
    }
    /** Get SSH status. */
    async status(workspaceId) {
        return this._http.request({ method: 'GET', path: `/workspace/${workspaceId}/ssh` });
    }
    /** Enable SSH. */
    async enable(workspaceId) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/ssh/enable` });
    }
    /** Disable SSH. */
    async disable(workspaceId) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/ssh/disable` });
    }
    /** Set SSH password. */
    async setPassword(workspaceId, params) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/ssh/password`, body: params });
    }
    /** Set authorized SSH key. */
    async setKey(workspaceId, params) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/ssh/key`, body: params });
    }
}
