// oblien/dist/resources/usage.js
/** Per-workspace usage tracking and account-level usage. */
export class UsageResource {
    _http;
    constructor(_http) {
        this._http = _http;
    }
    /** Get workspace usage. */
    async get(workspaceId, params = {}) {
        return this._http.request({
            method: 'GET',
            path: `/workspace/${workspaceId}/usage`,
            query: params,
        });
    }
    /** Get usage totals. */
    async totals(workspaceId) {
        return this._http.request({ method: 'GET', path: `/workspace/${workspaceId}/usage/totals` });
    }
    /** Get credits chart data. */
    async creditsChart(workspaceId) {
        return this._http.request({ method: 'GET', path: `/workspace/${workspaceId}/usage/credits/chart` });
    }
    /** Get global usage across all workspaces. */
    async global() {
        return this._http.request({ method: 'GET', path: '/workspace/usage' });
    }
    /** Get activity history. */
    async activity() {
        return this._http.request({ method: 'GET', path: '/workspace/activity' });
    }
    /** Wipe all usage data for a workspace. */
    async wipe(workspaceId) {
        return this._http.request({ method: 'DELETE', path: `/workspace/${workspaceId}/usage` });
    }
    /** Start usage tracking for a workspace. */
    async startTracking(workspaceId) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/usage/tracking/start` });
    }
    /** Stop usage tracking for a workspace. */
    async stopTracking(workspaceId) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/usage/tracking/stop` });
    }
}
