// oblien/dist/resources/workloads.js
/** Managed workloads (long-running processes). */
export class WorkloadsResource {
    _http;
    constructor(_http) {
        this._http = _http;
    }
    /** Create a workload. */
    async create(workspaceId, params) {
        const res = await this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/workloads`, body: params });
        return res.workload;
    }
    /** List workloads. */
    async list(workspaceId, params = {}) {
        const res = await this._http.request({
            method: 'GET',
            path: `/workspace/${workspaceId}/workloads`,
            query: params,
        });
        return res.workloads ?? [];
    }
    /** Get a workload. */
    async get(workspaceId, workloadId) {
        return this._http.request({ method: 'GET', path: `/workspace/${workspaceId}/workloads/${workloadId}` });
    }
    /** Update a workload. */
    async update(workspaceId, workloadId, params) {
        return this._http.request({ method: 'PUT', path: `/workspace/${workspaceId}/workloads/${workloadId}`, body: params });
    }
    /** Partial update a workload (PATCH alias). */
    async patch(workspaceId, workloadId, params) {
        return this._http.request({ method: 'PATCH', path: `/workspace/${workspaceId}/workloads/${workloadId}`, body: params });
    }
    /** Delete a workload. */
    async delete(workspaceId, workloadId) {
        return this._http.request({ method: 'DELETE', path: `/workspace/${workspaceId}/workloads/${workloadId}` });
    }
    /** Delete all workloads. */
    async deleteAll(workspaceId) {
        return this._http.request({ method: 'DELETE', path: `/workspace/${workspaceId}/workloads` });
    }
    /** Start a workload. */
    async start(workspaceId, workloadId) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/workloads/${workloadId}/start` });
    }
    /** Stop a workload. */
    async stop(workspaceId, workloadId) {
        return this._http.request({ method: 'POST', path: `/workspace/${workspaceId}/workloads/${workloadId}/stop` });
    }
    /** Get workload status. */
    async status(workspaceId, workloadId) {
        return this._http.request({ method: 'GET', path: `/workspace/${workspaceId}/workloads/${workloadId}/status` });
    }
    /** Get workload logs. */
    async logs(workspaceId, workloadId) {
        return this._http.request({ method: 'GET', path: `/workspace/${workspaceId}/workloads/${workloadId}/logs` });
    }
    /** Get workload stats. */
    async stats(workspaceId, workloadId) {
        return this._http.request({ method: 'GET', path: `/workspace/${workspaceId}/workloads/${workloadId}/stats` });
    }
    /** Stream workload stats (SSE). */
    async *statsStream(workspaceId, workloadId) {
        yield* this._http.requestSSE({
            method: 'GET',
            path: `/workspace/${workspaceId}/workloads/${workloadId}/stats/stream`,
        });
    }
    /** Stream workload logs (SSE). */
    async *logsStream(workspaceId, workloadId) {
        yield* this._http.requestSSE({
            method: 'GET',
            path: `/workspace/${workspaceId}/workloads/${workloadId}/logs/stream`,
        });
    }
    /** Stream stats for ALL workloads (SSE). */
    async *allStatsStream(workspaceId) {
        yield* this._http.requestSSE({
            method: 'GET',
            path: `/workspace/${workspaceId}/workloads/stats/stream`,
        });
    }
}
