// oblien/dist/error.js
/**
 * Structured error hierarchy for the Oblien SDK.
 * Every API error is wrapped in one of these classes so callers can
 * catch by type (`catch (e) { if (e instanceof NotFoundError) … }`).
 */
export class OblienError extends Error {
    status;
    code;
    constructor(message, status, code) {
        super(message);
        this.name = 'OblienError';
        this.status = status;
        this.code = code;
    }
}
export class AuthenticationError extends OblienError {
    constructor(message = 'Invalid or missing credentials') {
        super(message, 401, 'authentication_error');
        this.name = 'AuthenticationError';
    }
}
export class NotFoundError extends OblienError {
    constructor(message = 'Resource not found') {
        super(message, 404, 'not_found');
        this.name = 'NotFoundError';
    }
}
export class RateLimitError extends OblienError {
    constructor(message = 'Rate limit exceeded') {
        super(message, 429, 'rate_limited');
        this.name = 'RateLimitError';
    }
}
export class ValidationError extends OblienError {
    constructor(message = 'Validation failed') {
        super(message, 422, 'validation_error');
        this.name = 'ValidationError';
    }
}
export class PaymentRequiredError extends OblienError {
    constructor(message = 'Quota exceeded') {
        super(message, 402, 'payment_required');
        this.name = 'PaymentRequiredError';
    }
}
export class ConflictError extends OblienError {
    constructor(message = 'Resource in wrong state') {
        super(message, 409, 'conflict');
        this.name = 'ConflictError';
    }
}
