// oblien/dist/http.js
import { OblienError, AuthenticationError, NotFoundError, RateLimitError, ValidationError, PaymentRequiredError, ConflictError, } from './error.js';
/**
 * Low-level HTTP transport.
 * Uses native `fetch` (Node 18+). All auth headers are injected automatically.
 */
export class HttpClient {
    baseUrl;
    headers;
    constructor(options) {
        this.baseUrl = (options.baseUrl ?? 'https://api.oblien.com').replace(/\/$/, '');
        this.headers = {
            'Content-Type': 'application/json',
            'X-Client-ID': options.clientId,
            'X-Client-Secret': options.clientSecret,
        };
    }
    /** Execute an HTTP request and return the parsed JSON body. */
    async request(options) {
        const url = this.buildUrl(options.path, options.query);
        const init = {
            method: options.method,
            headers: { ...this.headers, ...options.headers },
        };
        if (options.body !== undefined) {
            init.body = JSON.stringify(options.body);
        }
        const res = await fetch(url, init);
        const body = await res.json().catch(() => null);
        if (!res.ok) {
            throw this.errorForStatus(res.status, body);
        }
        return body;
    }
    /**
     * Execute an SSE streaming request and yield parsed JSON events.
     * Used for stats streams, log streams, and other Server-Sent Events endpoints.
     */
    async *requestSSE(options) {
        const url = this.buildUrl(options.path, options.query);
        const init = {
            method: options.method,
            headers: { ...this.headers, ...options.headers, 'Accept': 'text/event-stream' },
        };
        if (options.body !== undefined) {
            init.body = JSON.stringify(options.body);
        }
        const res = await fetch(url, init);
        if (!res.ok || !res.body) {
            const body = await res.json().catch(() => null);
            throw this.errorForStatus(res.status, body);
        }
        const reader = res.body.getReader();
        const decoder = new TextDecoder();
        let buffer = '';
        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done)
                    break;
                buffer += decoder.decode(value, { stream: true });
                const lines = buffer.split('\n');
                buffer = lines.pop() ?? '';
                for (const line of lines) {
                    if (!line.startsWith('data: '))
                        continue;
                    const data = line.slice(6).trim();
                    if (!data || data === '[DONE]')
                        continue;
                    try {
                        yield JSON.parse(data);
                    }
                    catch {
                        // Skip non-JSON SSE data lines
                    }
                }
            }
        }
        finally {
            reader.releaseLock();
        }
    }
    // ── Internals ───────────────────────────────────────────────────────
    buildUrl(path, query) {
        const normalized = path.startsWith('/') ? path : `/${path}`;
        const url = new URL(normalized, this.baseUrl);
        if (query) {
            for (const [k, v] of Object.entries(query)) {
                if (v !== undefined)
                    url.searchParams.set(k, String(v));
            }
        }
        return url.toString();
    }
    errorForStatus(status, body) {
        const msg = body?.error ??
            `Request failed with status ${status}`;
        switch (status) {
            case 401: return new AuthenticationError(msg);
            case 402: return new PaymentRequiredError(msg);
            case 404: return new NotFoundError(msg);
            case 409: return new ConflictError(msg);
            case 422: return new ValidationError(msg);
            case 429: return new RateLimitError(msg);
            default: return new OblienError(msg, status);
        }
    }
}
