// oblien/dist/runtime-http.js
import { OblienError, AuthenticationError, NotFoundError, RateLimitError, ValidationError, ConflictError, } from './error.js';
/**
 * HTTP transport for the Runtime (workspace internal API).
 * Uses Bearer token auth against workspace.oblien.com or a direct IP.
 */
export class RuntimeHttpClient {
    baseUrl;
    headers;
    constructor(options) {
        this.baseUrl = (options.baseUrl ?? 'https://workspace.oblien.com').replace(/\/$/, '');
        this.headers = {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${options.token}`,
        };
    }
    async request(options) {
        const url = this.buildUrl(options.path, options.query);
        const init = {
            method: options.method,
            headers: { ...this.headers, ...options.headers },
        };
        if (options.body !== undefined) {
            init.body = JSON.stringify(options.body);
        }
        const res = await fetch(url, init);
        const body = await res.json().catch(() => null);
        if (!res.ok) {
            throw this.errorForStatus(res.status, body);
        }
        return body;
    }
    /** Send raw bytes (for stdin). */
    async requestRaw(path, data) {
        const url = this.buildUrl(path);
        const res = await fetch(url, {
            method: 'POST',
            headers: {
                ...this.headers,
                'Content-Type': 'application/octet-stream',
            },
            body: data,
        });
        const json = await res.json().catch(() => null);
        if (!res.ok)
            throw this.errorForStatus(res.status, json);
        return json;
    }
    /**
     * Get a raw streaming Response for SSE / NDJSON consumption.
     * Handles auth and error checking; callers parse the stream format.
     */
    async requestStream(options) {
        const url = this.buildUrl(options.path, options.query);
        const init = {
            method: options.method,
            headers: { ...this.headers, ...options.headers },
        };
        if (options.body !== undefined) {
            init.body = JSON.stringify(options.body);
        }
        const res = await fetch(url, init);
        if (!res.ok) {
            const body = await res.json().catch(() => null);
            throw this.errorForStatus(res.status, body);
        }
        return res;
    }
    getBaseUrl() {
        return this.baseUrl;
    }
    getHeaders() {
        return { ...this.headers };
    }
    // ── Internals ───────────────────────────────────────────────────────
    buildUrl(path, query) {
        const normalized = path.startsWith('/') ? path : `/${path}`;
        const url = new URL(normalized, this.baseUrl);
        if (query) {
            for (const [k, v] of Object.entries(query)) {
                if (v !== undefined)
                    url.searchParams.set(k, String(v));
            }
        }
        return url.toString();
    }
    errorForStatus(status, body) {
        const msg = body?.error ??
            `Request failed with status ${status}`;
        switch (status) {
            case 401: return new AuthenticationError(msg);
            case 404: return new NotFoundError(msg);
            case 409: return new ConflictError(msg);
            case 422: return new ValidationError(msg);
            case 429: return new RateLimitError(msg);
            default: return new OblienError(msg, status);
        }
    }
}
