// oblien/dist/runtime.js
import { RuntimeHttpClient } from './runtime-http.js';
import { FilesResource } from './runtime/files.js';
import { SearchResource } from './runtime/search.js';
import { ExecResource } from './runtime/exec.js';
import { TerminalResource } from './runtime/terminal.js';
import { WatcherResource } from './runtime/watcher.js';
import { RuntimeWebSocket } from './runtime/ws.js';
/**
 * Workspace Runtime — the data-plane client for interacting with a
 * running workspace's filesystem, terminal, commands, and file watchers.
 *
 * **Standalone** (direct token):
 * ```ts
 * import { Runtime } from 'oblien/runtime';
 *
 * const rt = new Runtime({ token: 'gateway_jwt_or_raw_token' });
 * // or with a direct IP
 * const rt = new Runtime({ token: 'raw_token', baseUrl: 'http://10.0.1.5:9990' });
 *
 * await rt.files.list({ dirPath: '/app' });
 * await rt.exec.run(['npm', 'test']);
 * ```
 *
 * **Via the Oblien client** (auto-fetches + caches JWT):
 * ```ts
 * import Oblien from 'oblien';
 *
 * const client = new Oblien({ clientId, clientSecret });
 * const rt = await client.workspaces.runtime('ws_abc');
 *
 * await rt.files.read({ filePath: '/app/index.js' });
 * await rt.terminal.create({ shell: '/bin/bash' });
 * ```
 */
export class Runtime {
    /** @internal */
    _http;
    /** @internal */
    _options;
    /** File operations — list, read, write, stream, stat, mkdir, delete. */
    files;
    /** Content search (ripgrep) and filename search. */
    search;
    /** Command execution — run, list, kill, stream, stdin. */
    exec;
    /** Interactive terminal sessions — create, list, close, scrollback. */
    terminal;
    /** File watchers — create, list, delete. */
    watcher;
    constructor(options) {
        this._options = options;
        this._http = new RuntimeHttpClient(options);
        this.files = new FilesResource(this._http);
        this.search = new SearchResource(this._http);
        this.exec = new ExecResource(this._http);
        this.terminal = new TerminalResource(this._http);
        this.watcher = new WatcherResource(this._http);
    }
    /** Health check — returns true if the runtime server is responding. */
    async health() {
        try {
            const res = await this._http.request({
                method: 'GET',
                path: '/health',
            });
            return res.status === 'ok';
        }
        catch {
            return false;
        }
    }
    /**
     * Open a multiplexed WebSocket connection for real-time terminal I/O
     * and file watcher events.
     *
     * ```ts
     * const ws = rt.ws();
     *
     * // Terminal I/O
     * ws.onTerminalOutput((id, data) => process.stdout.write(data));
     * ws.onTerminalExit((id, code) => console.log(`Terminal ${id} exited: ${code}`));
     * ws.writeTerminalInput(terminalId, 'ls -la\n');
     * ws.resizeTerminal(terminalId, 120, 40);
     *
     * // Watcher events
     * ws.onWatcherEvent((event) => console.log(event.path, event.op));
     *
     * // Lifecycle
     * ws.connect();
     * ws.close();
     * ```
     */
    ws(options) {
        return new RuntimeWebSocket(this._options, options);
    }
}
