// oblien/dist/workspace.js
import { Runtime } from './runtime.js';
import { LifecycleResource } from './resources/lifecycle.js';
import { NetworkResource } from './resources/network.js';
import { SshResource } from './resources/ssh.js';
import { PublicAccessResource } from './resources/public-access.js';
import { ResourcesResource } from './resources/resources.js';
import { SnapshotsResource } from './resources/snapshots.js';
import { WorkloadsResource } from './resources/workloads.js';
import { MetricsResource } from './resources/metrics.js';
import { UsageResource } from './resources/usage.js';
import { MetadataResource } from './resources/metadata.js';
import { ApiAccessResource } from './resources/api-access.js';
import { LogsResource } from './resources/logs.js';
import { ImagesResource } from './resources/images.js';
/**
 * Workspace operations namespace.
 *
 * Can be obtained two ways — both share the same class:
 *
 * ```ts
 * // Via the client (pre-wired)
 * import Oblien from 'oblien';
 * const client = new Oblien({ clientId, clientSecret });
 * await client.workspaces.create({ image: 'node' });
 *
 * // Standalone
 * import { Workspace } from 'oblien/workspace';
 * const ws = new Workspace(client);
 * await ws.create({ image: 'node' });
 * ```
 */
export class Workspace {
    /** @internal */
    _http;
    _base = '/workspace';
    /** Cached Runtime instances per workspace ID. */
    _runtimes = new Map();
    // ── Sub-resource namespaces ───────────────────────────────────────
    /** Lifecycle — permanent/temporary mode, TTL, ping. */
    lifecycle;
    /** Network — firewall, private links, public access, outbound IP. */
    network;
    /** SSH access — enable, disable, set password/key. */
    ssh;
    /** Public port access — expose and revoke ports. */
    publicAccess;
    /** Resources — CPU, memory, disk allocation. */
    resources;
    /** Snapshots and archives. */
    snapshots;
    /** Managed workloads (long-running processes). */
    workloads;
    /** Monitoring — live stats, VM info/config. */
    metrics;
    /** Usage tracking — per-workspace and global. */
    usage;
    /** Metadata key-values. */
    metadata;
    /** Internal API server management — enable, disable, tokens. */
    apiAccess;
    /** Logs — boot logs, command logs. */
    logs;
    /** Image catalog. */
    images;
    /**
     * @param client - An `Oblien` instance (or any object exposing `_http`).
     */
    constructor(client) {
        this._http = client._http;
        this.lifecycle = new LifecycleResource(this._http);
        this.network = new NetworkResource(this._http);
        this.ssh = new SshResource(this._http);
        this.publicAccess = new PublicAccessResource(this._http);
        this.resources = new ResourcesResource(this._http);
        this.snapshots = new SnapshotsResource(this._http);
        this.workloads = new WorkloadsResource(this._http);
        this.metrics = new MetricsResource(this._http);
        this.usage = new UsageResource(this._http);
        this.metadata = new MetadataResource(this._http);
        this.apiAccess = new ApiAccessResource(this._http);
        this.logs = new LogsResource(this._http);
        this.images = new ImagesResource(this._http);
    }
    // ── CRUD ──────────────────────────────────────────────────────────
    /** Create a new workspace. */
    async create(params = {}) {
        const res = await this._http.request({ method: 'POST', path: this._base, body: params });
        return res.workspace;
    }
    /** List workspaces with optional pagination / filters. */
    async list(params = {}) {
        return this._http.request({
            method: 'GET',
            path: this._base,
            query: params,
        });
    }
    /** Get a single workspace by ID. */
    async get(workspaceId) {
        const res = await this._http.request({ method: 'GET', path: `${this._base}/${workspaceId}` });
        return res.workspace;
    }
    /** Update workspace settings. */
    async update(workspaceId, params) {
        const res = await this._http.request({ method: 'PUT', path: `${this._base}/${workspaceId}`, body: params });
        return res.workspace;
    }
    /** Permanently delete a workspace. */
    async delete(workspaceId) {
        return this._http.request({ method: 'DELETE', path: `${this._base}/${workspaceId}` });
    }
    /** Get aggregated workspace details. */
    async getDetails(workspaceId) {
        return this._http.request({ method: 'GET', path: `${this._base}/${workspaceId}/details` });
    }
    // ── Power / lifecycle shortcuts ───────────────────────────────────
    /** Start a workspace. */
    async start(workspaceId, params) {
        return this._http.request({ method: 'POST', path: `${this._base}/${workspaceId}/start`, body: params });
    }
    /** Stop a workspace. */
    async stop(workspaceId) {
        return this._http.request({ method: 'POST', path: `${this._base}/${workspaceId}/stop` });
    }
    /** Restart a workspace. */
    async restart(workspaceId) {
        return this._http.request({ method: 'POST', path: `${this._base}/${workspaceId}/restart` });
    }
    /** Pause a workspace (preserves memory state). */
    async pause(workspaceId) {
        return this._http.request({ method: 'POST', path: `${this._base}/${workspaceId}/pause` });
    }
    /** Resume a paused workspace. */
    async resume(workspaceId) {
        return this._http.request({ method: 'POST', path: `${this._base}/${workspaceId}/resume` });
    }
    /** Keep-alive ping — renews TTL for temporary workspaces. */
    async ping(workspaceId, params) {
        return this._http.request({ method: 'POST', path: `${this._base}/${workspaceId}/ping`, body: params });
    }
    // ── Account-level queries ─────────────────────────────────────────
    /** Get resource quota and limits. */
    async getQuota() {
        return this._http.request({ method: 'GET', path: `${this._base}/quota` });
    }
    /** Estimate monthly credit usage. */
    async getEstimate(params) {
        return this._http.request({
            method: 'GET',
            path: `${this._base}/estimate`,
            query: params,
        });
    }
    // ── Runtime (data plane) ──────────────────────────────────────────
    /**
     * Get a {@link Runtime} handle for a workspace.
     *
     * On first call, enables the internal API server (if needed) and
     * caches the Gateway JWT. Subsequent calls for the same workspace ID
     * return the cached Runtime instantly.
     *
     * Pass `{ force: true }` to bypass the cache and fetch a fresh token
     * (useful when a cached JWT has expired).
     *
     * ```ts
     * const rt = await client.workspaces.runtime('ws_abc');
     * await rt.files.list({ dirPath: '/app' });
     * await rt.exec.run(['npm', 'test']);
     * ```
     */
    async runtime(workspaceId, options) {
        if (!options?.force) {
            const cached = this._runtimes.get(workspaceId);
            if (cached)
                return cached;
        }
        const access = await this.apiAccess.enable(workspaceId);
        const rt = new Runtime({ token: access.token });
        this._runtimes.set(workspaceId, rt);
        return rt;
    }
    /**
     * Invalidate the cached Runtime for a workspace.
     * Next call to `runtime()` will fetch a fresh token.
     */
    invalidateRuntime(workspaceId) {
        this._runtimes.delete(workspaceId);
    }
    /** Clear all cached Runtime instances. */
    invalidateAllRuntimes() {
        this._runtimes.clear();
    }
}
